// mdi-material-ui/light/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Account", {
  enumerable: true,
  get: function () {
    return _Account.default;
  }
});
Object.defineProperty(exports, "Alarm", {
  enumerable: true,
  get: function () {
    return _Alarm.default;
  }
});
Object.defineProperty(exports, "AlarmPlus", {
  enumerable: true,
  get: function () {
    return _AlarmPlus.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function () {
    return _Alert.default;
  }
});
Object.defineProperty(exports, "AlertCircle", {
  enumerable: true,
  get: function () {
    return _AlertCircle.default;
  }
});
Object.defineProperty(exports, "AlertOctagon", {
  enumerable: true,
  get: function () {
    return _AlertOctagon.default;
  }
});
Object.defineProperty(exports, "ArrangeBringForward", {
  enumerable: true,
  get: function () {
    return _ArrangeBringForward.default;
  }
});
Object.defineProperty(exports, "ArrangeBringToFront", {
  enumerable: true,
  get: function () {
    return _ArrangeBringToFront.default;
  }
});
Object.defineProperty(exports, "ArrangeSendBackward", {
  enumerable: true,
  get: function () {
    return _ArrangeSendBackward.default;
  }
});
Object.defineProperty(exports, "ArrangeSendToBack", {
  enumerable: true,
  get: function () {
    return _ArrangeSendToBack.default;
  }
});
Object.defineProperty(exports, "ArrowDown", {
  enumerable: true,
  get: function () {
    return _ArrowDown.default;
  }
});
Object.defineProperty(exports, "ArrowDownCircle", {
  enumerable: true,
  get: function () {
    return _ArrowDownCircle.default;
  }
});
Object.defineProperty(exports, "ArrowLeft", {
  enumerable: true,
  get: function () {
    return _ArrowLeft.default;
  }
});
Object.defineProperty(exports, "ArrowLeftCircle", {
  enumerable: true,
  get: function () {
    return _ArrowLeftCircle.default;
  }
});
Object.defineProperty(exports, "ArrowRight", {
  enumerable: true,
  get: function () {
    return _ArrowRight.default;
  }
});
Object.defineProperty(exports, "ArrowRightCircle", {
  enumerable: true,
  get: function () {
    return _ArrowRightCircle.default;
  }
});
Object.defineProperty(exports, "ArrowUp", {
  enumerable: true,
  get: function () {
    return _ArrowUp.default;
  }
});
Object.defineProperty(exports, "ArrowUpCircle", {
  enumerable: true,
  get: function () {
    return _ArrowUpCircle.default;
  }
});
Object.defineProperty(exports, "Bank", {
  enumerable: true,
  get: function () {
    return _Bank.default;
  }
});
Object.defineProperty(exports, "Bell", {
  enumerable: true,
  get: function () {
    return _Bell.default;
  }
});
Object.defineProperty(exports, "BellOff", {
  enumerable: true,
  get: function () {
    return _BellOff.default;
  }
});
Object.defineProperty(exports, "Bluetooth", {
  enumerable: true,
  get: function () {
    return _Bluetooth.default;
  }
});
Object.defineProperty(exports, "Book", {
  enumerable: true,
  get: function () {
    return _Book.default;
  }
});
Object.defineProperty(exports, "BookMultiple", {
  enumerable: true,
  get: function () {
    return _BookMultiple.default;
  }
});
Object.defineProperty(exports, "BookPlus", {
  enumerable: true,
  get: function () {
    return _BookPlus.default;
  }
});
Object.defineProperty(exports, "Bookmark", {
  enumerable: true,
  get: function () {
    return _Bookmark.default;
  }
});
Object.defineProperty(exports, "BorderAll", {
  enumerable: true,
  get: function () {
    return _BorderAll.default;
  }
});
Object.defineProperty(exports, "BorderBottom", {
  enumerable: true,
  get: function () {
    return _BorderBottom.default;
  }
});
Object.defineProperty(exports, "BorderHorizontal", {
  enumerable: true,
  get: function () {
    return _BorderHorizontal.default;
  }
});
Object.defineProperty(exports, "BorderInside", {
  enumerable: true,
  get: function () {
    return _BorderInside.default;
  }
});
Object.defineProperty(exports, "BorderLeft", {
  enumerable: true,
  get: function () {
    return _BorderLeft.default;
  }
});
Object.defineProperty(exports, "BorderNone", {
  enumerable: true,
  get: function () {
    return _BorderNone.default;
  }
});
Object.defineProperty(exports, "BorderOutside", {
  enumerable: true,
  get: function () {
    return _BorderOutside.default;
  }
});
Object.defineProperty(exports, "BorderRight", {
  enumerable: true,
  get: function () {
    return _BorderRight.default;
  }
});
Object.defineProperty(exports, "BorderTop", {
  enumerable: true,
  get: function () {
    return _BorderTop.default;
  }
});
Object.defineProperty(exports, "BorderVertical", {
  enumerable: true,
  get: function () {
    return _BorderVertical.default;
  }
});
Object.defineProperty(exports, "Briefcase", {
  enumerable: true,
  get: function () {
    return _Briefcase.default;
  }
});
Object.defineProperty(exports, "Calendar", {
  enumerable: true,
  get: function () {
    return _Calendar.default;
  }
});
Object.defineProperty(exports, "Camcorder", {
  enumerable: true,
  get: function () {
    return _Camcorder.default;
  }
});
Object.defineProperty(exports, "Camera", {
  enumerable: true,
  get: function () {
    return _Camera.default;
  }
});
Object.defineProperty(exports, "Cancel", {
  enumerable: true,
  get: function () {
    return _Cancel.default;
  }
});
Object.defineProperty(exports, "Cart", {
  enumerable: true,
  get: function () {
    return _Cart.default;
  }
});
Object.defineProperty(exports, "ChartAreaspline", {
  enumerable: true,
  get: function () {
    return _ChartAreaspline.default;
  }
});
Object.defineProperty(exports, "ChartBar", {
  enumerable: true,
  get: function () {
    return _ChartBar.default;
  }
});
Object.defineProperty(exports, "ChartHistogram", {
  enumerable: true,
  get: function () {
    return _ChartHistogram.default;
  }
});
Object.defineProperty(exports, "ChartLine", {
  enumerable: true,
  get: function () {
    return _ChartLine.default;
  }
});
Object.defineProperty(exports, "ChartPie", {
  enumerable: true,
  get: function () {
    return _ChartPie.default;
  }
});
Object.defineProperty(exports, "Check", {
  enumerable: true,
  get: function () {
    return _Check.default;
  }
});
Object.defineProperty(exports, "ChevronDoubleDown", {
  enumerable: true,
  get: function () {
    return _ChevronDoubleDown.default;
  }
});
Object.defineProperty(exports, "ChevronDoubleLeft", {
  enumerable: true,
  get: function () {
    return _ChevronDoubleLeft.default;
  }
});
Object.defineProperty(exports, "ChevronDoubleRight", {
  enumerable: true,
  get: function () {
    return _ChevronDoubleRight.default;
  }
});
Object.defineProperty(exports, "ChevronDoubleUp", {
  enumerable: true,
  get: function () {
    return _ChevronDoubleUp.default;
  }
});
Object.defineProperty(exports, "ChevronDown", {
  enumerable: true,
  get: function () {
    return _ChevronDown.default;
  }
});
Object.defineProperty(exports, "ChevronLeft", {
  enumerable: true,
  get: function () {
    return _ChevronLeft.default;
  }
});
Object.defineProperty(exports, "ChevronRight", {
  enumerable: true,
  get: function () {
    return _ChevronRight.default;
  }
});
Object.defineProperty(exports, "ChevronUp", {
  enumerable: true,
  get: function () {
    return _ChevronUp.default;
  }
});
Object.defineProperty(exports, "Circle", {
  enumerable: true,
  get: function () {
    return _Circle.default;
  }
});
Object.defineProperty(exports, "Clipboard", {
  enumerable: true,
  get: function () {
    return _Clipboard.default;
  }
});
Object.defineProperty(exports, "ClipboardCheck", {
  enumerable: true,
  get: function () {
    return _ClipboardCheck.default;
  }
});
Object.defineProperty(exports, "ClipboardPlus", {
  enumerable: true,
  get: function () {
    return _ClipboardPlus.default;
  }
});
Object.defineProperty(exports, "ClipboardText", {
  enumerable: true,
  get: function () {
    return _ClipboardText.default;
  }
});
Object.defineProperty(exports, "Clock", {
  enumerable: true,
  get: function () {
    return _Clock.default;
  }
});
Object.defineProperty(exports, "ClosedCaption", {
  enumerable: true,
  get: function () {
    return _ClosedCaption.default;
  }
});
Object.defineProperty(exports, "Cloud", {
  enumerable: true,
  get: function () {
    return _Cloud.default;
  }
});
Object.defineProperty(exports, "CloudDownload", {
  enumerable: true,
  get: function () {
    return _CloudDownload.default;
  }
});
Object.defineProperty(exports, "CloudUpload", {
  enumerable: true,
  get: function () {
    return _CloudUpload.default;
  }
});
Object.defineProperty(exports, "Comment", {
  enumerable: true,
  get: function () {
    return _Comment.default;
  }
});
Object.defineProperty(exports, "CommentAlert", {
  enumerable: true,
  get: function () {
    return _CommentAlert.default;
  }
});
Object.defineProperty(exports, "CommentText", {
  enumerable: true,
  get: function () {
    return _CommentText.default;
  }
});
Object.defineProperty(exports, "Console", {
  enumerable: true,
  get: function () {
    return _Console.default;
  }
});
Object.defineProperty(exports, "ContentCut", {
  enumerable: true,
  get: function () {
    return _ContentCut.default;
  }
});
Object.defineProperty(exports, "ContentDuplicate", {
  enumerable: true,
  get: function () {
    return _ContentDuplicate.default;
  }
});
Object.defineProperty(exports, "ContentPaste", {
  enumerable: true,
  get: function () {
    return _ContentPaste.default;
  }
});
Object.defineProperty(exports, "ContentSave", {
  enumerable: true,
  get: function () {
    return _ContentSave.default;
  }
});
Object.defineProperty(exports, "ContentSaveAll", {
  enumerable: true,
  get: function () {
    return _ContentSaveAll.default;
  }
});
Object.defineProperty(exports, "CreditCard", {
  enumerable: true,
  get: function () {
    return _CreditCard.default;
  }
});
Object.defineProperty(exports, "CreditCardScan", {
  enumerable: true,
  get: function () {
    return _CreditCardScan.default;
  }
});
Object.defineProperty(exports, "Crop", {
  enumerable: true,
  get: function () {
    return _Crop.default;
  }
});
Object.defineProperty(exports, "CropFree", {
  enumerable: true,
  get: function () {
    return _CropFree.default;
  }
});
Object.defineProperty(exports, "CurrencyEur", {
  enumerable: true,
  get: function () {
    return _CurrencyEur.default;
  }
});
Object.defineProperty(exports, "CurrencyGbp", {
  enumerable: true,
  get: function () {
    return _CurrencyGbp.default;
  }
});
Object.defineProperty(exports, "CurrencyRub", {
  enumerable: true,
  get: function () {
    return _CurrencyRub.default;
  }
});
Object.defineProperty(exports, "CurrencyUsd", {
  enumerable: true,
  get: function () {
    return _CurrencyUsd.default;
  }
});
Object.defineProperty(exports, "Delete", {
  enumerable: true,
  get: function () {
    return _Delete.default;
  }
});
Object.defineProperty(exports, "DiamondStone", {
  enumerable: true,
  get: function () {
    return _DiamondStone.default;
  }
});
Object.defineProperty(exports, "DotsHorizontal", {
  enumerable: true,
  get: function () {
    return _DotsHorizontal.default;
  }
});
Object.defineProperty(exports, "DotsVertical", {
  enumerable: true,
  get: function () {
    return _DotsVertical.default;
  }
});
Object.defineProperty(exports, "Download", {
  enumerable: true,
  get: function () {
    return _Download.default;
  }
});
Object.defineProperty(exports, "Eject", {
  enumerable: true,
  get: function () {
    return _Eject.default;
  }
});
Object.defineProperty(exports, "Email", {
  enumerable: true,
  get: function () {
    return _Email.default;
  }
});
Object.defineProperty(exports, "EmailOpen", {
  enumerable: true,
  get: function () {
    return _EmailOpen.default;
  }
});
Object.defineProperty(exports, "Eye", {
  enumerable: true,
  get: function () {
    return _Eye.default;
  }
});
Object.defineProperty(exports, "EyeOff", {
  enumerable: true,
  get: function () {
    return _EyeOff.default;
  }
});
Object.defineProperty(exports, "Factory", {
  enumerable: true,
  get: function () {
    return _Factory.default;
  }
});
Object.defineProperty(exports, "FastForward", {
  enumerable: true,
  get: function () {
    return _FastForward.default;
  }
});
Object.defineProperty(exports, "File", {
  enumerable: true,
  get: function () {
    return _File.default;
  }
});
Object.defineProperty(exports, "FileMultiple", {
  enumerable: true,
  get: function () {
    return _FileMultiple.default;
  }
});
Object.defineProperty(exports, "FilePlus", {
  enumerable: true,
  get: function () {
    return _FilePlus.default;
  }
});
Object.defineProperty(exports, "Flag", {
  enumerable: true,
  get: function () {
    return _Flag.default;
  }
});
Object.defineProperty(exports, "Flash", {
  enumerable: true,
  get: function () {
    return _Flash.default;
  }
});
Object.defineProperty(exports, "Flask", {
  enumerable: true,
  get: function () {
    return _Flask.default;
  }
});
Object.defineProperty(exports, "FlaskEmpty", {
  enumerable: true,
  get: function () {
    return _FlaskEmpty.default;
  }
});
Object.defineProperty(exports, "Folder", {
  enumerable: true,
  get: function () {
    return _Folder.default;
  }
});
Object.defineProperty(exports, "FolderMultiple", {
  enumerable: true,
  get: function () {
    return _FolderMultiple.default;
  }
});
Object.defineProperty(exports, "FolderPlus", {
  enumerable: true,
  get: function () {
    return _FolderPlus.default;
  }
});
Object.defineProperty(exports, "FormatAlignBottom", {
  enumerable: true,
  get: function () {
    return _FormatAlignBottom.default;
  }
});
Object.defineProperty(exports, "FormatAlignCenter", {
  enumerable: true,
  get: function () {
    return _FormatAlignCenter.default;
  }
});
Object.defineProperty(exports, "FormatAlignJustify", {
  enumerable: true,
  get: function () {
    return _FormatAlignJustify.default;
  }
});
Object.defineProperty(exports, "FormatAlignLeft", {
  enumerable: true,
  get: function () {
    return _FormatAlignLeft.default;
  }
});
Object.defineProperty(exports, "FormatAlignMiddle", {
  enumerable: true,
  get: function () {
    return _FormatAlignMiddle.default;
  }
});
Object.defineProperty(exports, "FormatAlignRight", {
  enumerable: true,
  get: function () {
    return _FormatAlignRight.default;
  }
});
Object.defineProperty(exports, "FormatAlignTop", {
  enumerable: true,
  get: function () {
    return _FormatAlignTop.default;
  }
});
Object.defineProperty(exports, "FormatBold", {
  enumerable: true,
  get: function () {
    return _FormatBold.default;
  }
});
Object.defineProperty(exports, "FormatClear", {
  enumerable: true,
  get: function () {
    return _FormatClear.default;
  }
});
Object.defineProperty(exports, "FormatFloatCenter", {
  enumerable: true,
  get: function () {
    return _FormatFloatCenter.default;
  }
});
Object.defineProperty(exports, "FormatFloatLeft", {
  enumerable: true,
  get: function () {
    return _FormatFloatLeft.default;
  }
});
Object.defineProperty(exports, "FormatFloatNone", {
  enumerable: true,
  get: function () {
    return _FormatFloatNone.default;
  }
});
Object.defineProperty(exports, "FormatFloatRight", {
  enumerable: true,
  get: function () {
    return _FormatFloatRight.default;
  }
});
Object.defineProperty(exports, "FormatIndentDecrease", {
  enumerable: true,
  get: function () {
    return _FormatIndentDecrease.default;
  }
});
Object.defineProperty(exports, "FormatIndentIncrease", {
  enumerable: true,
  get: function () {
    return _FormatIndentIncrease.default;
  }
});
Object.defineProperty(exports, "FormatItalic", {
  enumerable: true,
  get: function () {
    return _FormatItalic.default;
  }
});
Object.defineProperty(exports, "FormatLineSpacing", {
  enumerable: true,
  get: function () {
    return _FormatLineSpacing.default;
  }
});
Object.defineProperty(exports, "FormatListBulleted", {
  enumerable: true,
  get: function () {
    return _FormatListBulleted.default;
  }
});
Object.defineProperty(exports, "FormatListChecks", {
  enumerable: true,
  get: function () {
    return _FormatListChecks.default;
  }
});
Object.defineProperty(exports, "FormatListNumbers", {
  enumerable: true,
  get: function () {
    return _FormatListNumbers.default;
  }
});
Object.defineProperty(exports, "FormatQuoteClose", {
  enumerable: true,
  get: function () {
    return _FormatQuoteClose.default;
  }
});
Object.defineProperty(exports, "FormatQuoteOpen", {
  enumerable: true,
  get: function () {
    return _FormatQuoteOpen.default;
  }
});
Object.defineProperty(exports, "FormatUnderline", {
  enumerable: true,
  get: function () {
    return _FormatUnderline.default;
  }
});
Object.defineProperty(exports, "FormatWrapInline", {
  enumerable: true,
  get: function () {
    return _FormatWrapInline.default;
  }
});
Object.defineProperty(exports, "FormatWrapSquare", {
  enumerable: true,
  get: function () {
    return _FormatWrapSquare.default;
  }
});
Object.defineProperty(exports, "FormatWrapTight", {
  enumerable: true,
  get: function () {
    return _FormatWrapTight.default;
  }
});
Object.defineProperty(exports, "FormatWrapTopBottom", {
  enumerable: true,
  get: function () {
    return _FormatWrapTopBottom.default;
  }
});
Object.defineProperty(exports, "Forum", {
  enumerable: true,
  get: function () {
    return _Forum.default;
  }
});
Object.defineProperty(exports, "Fullscreen", {
  enumerable: true,
  get: function () {
    return _Fullscreen.default;
  }
});
Object.defineProperty(exports, "FullscreenExit", {
  enumerable: true,
  get: function () {
    return _FullscreenExit.default;
  }
});
Object.defineProperty(exports, "Gift", {
  enumerable: true,
  get: function () {
    return _Gift.default;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function () {
    return _Grid.default;
  }
});
Object.defineProperty(exports, "GridLarge", {
  enumerable: true,
  get: function () {
    return _GridLarge.default;
  }
});
Object.defineProperty(exports, "GridOff", {
  enumerable: true,
  get: function () {
    return _GridOff.default;
  }
});
Object.defineProperty(exports, "Group", {
  enumerable: true,
  get: function () {
    return _Group.default;
  }
});
Object.defineProperty(exports, "Hamburger", {
  enumerable: true,
  get: function () {
    return _Hamburger.default;
  }
});
Object.defineProperty(exports, "Heart", {
  enumerable: true,
  get: function () {
    return _Heart.default;
  }
});
Object.defineProperty(exports, "HeartHalf", {
  enumerable: true,
  get: function () {
    return _HeartHalf.default;
  }
});
Object.defineProperty(exports, "HeartOff", {
  enumerable: true,
  get: function () {
    return _HeartOff.default;
  }
});
Object.defineProperty(exports, "Help", {
  enumerable: true,
  get: function () {
    return _Help.default;
  }
});
Object.defineProperty(exports, "HelpCircle", {
  enumerable: true,
  get: function () {
    return _HelpCircle.default;
  }
});
Object.defineProperty(exports, "Hexagon", {
  enumerable: true,
  get: function () {
    return _Hexagon.default;
  }
});
Object.defineProperty(exports, "Home", {
  enumerable: true,
  get: function () {
    return _Home.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _Image.default;
  }
});
Object.defineProperty(exports, "Inbox", {
  enumerable: true,
  get: function () {
    return _Inbox.default;
  }
});
Object.defineProperty(exports, "Information", {
  enumerable: true,
  get: function () {
    return _Information.default;
  }
});
Object.defineProperty(exports, "Label", {
  enumerable: true,
  get: function () {
    return _Label.default;
  }
});
Object.defineProperty(exports, "Layers", {
  enumerable: true,
  get: function () {
    return _Layers.default;
  }
});
Object.defineProperty(exports, "Lightbulb", {
  enumerable: true,
  get: function () {
    return _Lightbulb.default;
  }
});
Object.defineProperty(exports, "LightbulbOn", {
  enumerable: true,
  get: function () {
    return _LightbulbOn.default;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function () {
    return _Link.default;
  }
});
Object.defineProperty(exports, "LinkVariant", {
  enumerable: true,
  get: function () {
    return _LinkVariant.default;
  }
});
Object.defineProperty(exports, "Lock", {
  enumerable: true,
  get: function () {
    return _Lock.default;
  }
});
Object.defineProperty(exports, "LockOpen", {
  enumerable: true,
  get: function () {
    return _LockOpen.default;
  }
});
Object.defineProperty(exports, "Login", {
  enumerable: true,
  get: function () {
    return _Login.default;
  }
});
Object.defineProperty(exports, "Logout", {
  enumerable: true,
  get: function () {
    return _Logout.default;
  }
});
Object.defineProperty(exports, "Magnify", {
  enumerable: true,
  get: function () {
    return _Magnify.default;
  }
});
Object.defineProperty(exports, "MagnifyMinus", {
  enumerable: true,
  get: function () {
    return _MagnifyMinus.default;
  }
});
Object.defineProperty(exports, "MagnifyPlus", {
  enumerable: true,
  get: function () {
    return _MagnifyPlus.default;
  }
});
Object.defineProperty(exports, "MapMarker", {
  enumerable: true,
  get: function () {
    return _MapMarker.default;
  }
});
Object.defineProperty(exports, "Memory", {
  enumerable: true,
  get: function () {
    return _Memory.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _Menu.default;
  }
});
Object.defineProperty(exports, "Message", {
  enumerable: true,
  get: function () {
    return _Message.default;
  }
});
Object.defineProperty(exports, "MessageAlert", {
  enumerable: true,
  get: function () {
    return _MessageAlert.default;
  }
});
Object.defineProperty(exports, "MessageImage", {
  enumerable: true,
  get: function () {
    return _MessageImage.default;
  }
});
Object.defineProperty(exports, "MessageProcessing", {
  enumerable: true,
  get: function () {
    return _MessageProcessing.default;
  }
});
Object.defineProperty(exports, "MessageReply", {
  enumerable: true,
  get: function () {
    return _MessageReply.default;
  }
});
Object.defineProperty(exports, "MessageText", {
  enumerable: true,
  get: function () {
    return _MessageText.default;
  }
});
Object.defineProperty(exports, "MessageVideo", {
  enumerable: true,
  get: function () {
    return _MessageVideo.default;
  }
});
Object.defineProperty(exports, "Microphone", {
  enumerable: true,
  get: function () {
    return _Microphone.default;
  }
});
Object.defineProperty(exports, "MicrophoneOff", {
  enumerable: true,
  get: function () {
    return _MicrophoneOff.default;
  }
});
Object.defineProperty(exports, "Minus", {
  enumerable: true,
  get: function () {
    return _Minus.default;
  }
});
Object.defineProperty(exports, "MinusBox", {
  enumerable: true,
  get: function () {
    return _MinusBox.default;
  }
});
Object.defineProperty(exports, "MinusCircle", {
  enumerable: true,
  get: function () {
    return _MinusCircle.default;
  }
});
Object.defineProperty(exports, "Monitor", {
  enumerable: true,
  get: function () {
    return _Monitor.default;
  }
});
Object.defineProperty(exports, "MonitorMultiple", {
  enumerable: true,
  get: function () {
    return _MonitorMultiple.default;
  }
});
Object.defineProperty(exports, "Music", {
  enumerable: true,
  get: function () {
    return _Music.default;
  }
});
Object.defineProperty(exports, "MusicOff", {
  enumerable: true,
  get: function () {
    return _MusicOff.default;
  }
});
Object.defineProperty(exports, "NfcVariant", {
  enumerable: true,
  get: function () {
    return _NfcVariant.default;
  }
});
Object.defineProperty(exports, "Note", {
  enumerable: true,
  get: function () {
    return _Note.default;
  }
});
Object.defineProperty(exports, "NoteMultiple", {
  enumerable: true,
  get: function () {
    return _NoteMultiple.default;
  }
});
Object.defineProperty(exports, "NotePlus", {
  enumerable: true,
  get: function () {
    return _NotePlus.default;
  }
});
Object.defineProperty(exports, "NoteText", {
  enumerable: true,
  get: function () {
    return _NoteText.default;
  }
});
Object.defineProperty(exports, "Octagon", {
  enumerable: true,
  get: function () {
    return _Octagon.default;
  }
});
Object.defineProperty(exports, "Paperclip", {
  enumerable: true,
  get: function () {
    return _Paperclip.default;
  }
});
Object.defineProperty(exports, "Pause", {
  enumerable: true,
  get: function () {
    return _Pause.default;
  }
});
Object.defineProperty(exports, "Pencil", {
  enumerable: true,
  get: function () {
    return _Pencil.default;
  }
});
Object.defineProperty(exports, "Phone", {
  enumerable: true,
  get: function () {
    return _Phone.default;
  }
});
Object.defineProperty(exports, "Pin", {
  enumerable: true,
  get: function () {
    return _Pin.default;
  }
});
Object.defineProperty(exports, "PinOff", {
  enumerable: true,
  get: function () {
    return _PinOff.default;
  }
});
Object.defineProperty(exports, "Play", {
  enumerable: true,
  get: function () {
    return _Play.default;
  }
});
Object.defineProperty(exports, "Plus", {
  enumerable: true,
  get: function () {
    return _Plus.default;
  }
});
Object.defineProperty(exports, "PlusBox", {
  enumerable: true,
  get: function () {
    return _PlusBox.default;
  }
});
Object.defineProperty(exports, "PlusCircle", {
  enumerable: true,
  get: function () {
    return _PlusCircle.default;
  }
});
Object.defineProperty(exports, "Power", {
  enumerable: true,
  get: function () {
    return _Power.default;
  }
});
Object.defineProperty(exports, "Presentation", {
  enumerable: true,
  get: function () {
    return _Presentation.default;
  }
});
Object.defineProperty(exports, "PresentationPlay", {
  enumerable: true,
  get: function () {
    return _PresentationPlay.default;
  }
});
Object.defineProperty(exports, "Printer", {
  enumerable: true,
  get: function () {
    return _Printer.default;
  }
});
Object.defineProperty(exports, "RedoVariant", {
  enumerable: true,
  get: function () {
    return _RedoVariant.default;
  }
});
Object.defineProperty(exports, "Refresh", {
  enumerable: true,
  get: function () {
    return _Refresh.default;
  }
});
Object.defineProperty(exports, "Repeat", {
  enumerable: true,
  get: function () {
    return _Repeat.default;
  }
});
Object.defineProperty(exports, "RepeatOff", {
  enumerable: true,
  get: function () {
    return _RepeatOff.default;
  }
});
Object.defineProperty(exports, "RepeatOnce", {
  enumerable: true,
  get: function () {
    return _RepeatOnce.default;
  }
});
Object.defineProperty(exports, "Rewind", {
  enumerable: true,
  get: function () {
    return _Rewind.default;
  }
});
Object.defineProperty(exports, "Rhombus", {
  enumerable: true,
  get: function () {
    return _Rhombus.default;
  }
});
Object.defineProperty(exports, "Rss", {
  enumerable: true,
  get: function () {
    return _Rss.default;
  }
});
Object.defineProperty(exports, "Script", {
  enumerable: true,
  get: function () {
    return _Script.default;
  }
});
Object.defineProperty(exports, "Settings", {
  enumerable: true,
  get: function () {
    return _Settings.default;
  }
});
Object.defineProperty(exports, "Shield", {
  enumerable: true,
  get: function () {
    return _Shield.default;
  }
});
Object.defineProperty(exports, "Shuffle", {
  enumerable: true,
  get: function () {
    return _Shuffle.default;
  }
});
Object.defineProperty(exports, "Signal", {
  enumerable: true,
  get: function () {
    return _Signal.default;
  }
});
Object.defineProperty(exports, "Sitemap", {
  enumerable: true,
  get: function () {
    return _Sitemap.default;
  }
});
Object.defineProperty(exports, "SkipNext", {
  enumerable: true,
  get: function () {
    return _SkipNext.default;
  }
});
Object.defineProperty(exports, "SkipPrevious", {
  enumerable: true,
  get: function () {
    return _SkipPrevious.default;
  }
});
Object.defineProperty(exports, "Sleep", {
  enumerable: true,
  get: function () {
    return _Sleep.default;
  }
});
Object.defineProperty(exports, "SleepOff", {
  enumerable: true,
  get: function () {
    return _SleepOff.default;
  }
});
Object.defineProperty(exports, "Spellcheck", {
  enumerable: true,
  get: function () {
    return _Spellcheck.default;
  }
});
Object.defineProperty(exports, "Square", {
  enumerable: true,
  get: function () {
    return _Square.default;
  }
});
Object.defineProperty(exports, "Star", {
  enumerable: true,
  get: function () {
    return _Star.default;
  }
});
Object.defineProperty(exports, "StarHalf", {
  enumerable: true,
  get: function () {
    return _StarHalf.default;
  }
});
Object.defineProperty(exports, "Stop", {
  enumerable: true,
  get: function () {
    return _Stop.default;
  }
});
Object.defineProperty(exports, "Tab", {
  enumerable: true,
  get: function () {
    return _Tab.default;
  }
});
Object.defineProperty(exports, "TabPlus", {
  enumerable: true,
  get: function () {
    return _TabPlus.default;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _Table.default;
  }
});
Object.defineProperty(exports, "Taco", {
  enumerable: true,
  get: function () {
    return _Taco.default;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function () {
    return _Tag.default;
  }
});
Object.defineProperty(exports, "Television", {
  enumerable: true,
  get: function () {
    return _Television.default;
  }
});
Object.defineProperty(exports, "ThumbDown", {
  enumerable: true,
  get: function () {
    return _ThumbDown.default;
  }
});
Object.defineProperty(exports, "ThumbUp", {
  enumerable: true,
  get: function () {
    return _ThumbUp.default;
  }
});
Object.defineProperty(exports, "ThumbsUpDown", {
  enumerable: true,
  get: function () {
    return _ThumbsUpDown.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function () {
    return _Tooltip.default;
  }
});
Object.defineProperty(exports, "TooltipText", {
  enumerable: true,
  get: function () {
    return _TooltipText.default;
  }
});
Object.defineProperty(exports, "Triangle", {
  enumerable: true,
  get: function () {
    return _Triangle.default;
  }
});
Object.defineProperty(exports, "Trophy", {
  enumerable: true,
  get: function () {
    return _Trophy.default;
  }
});
Object.defineProperty(exports, "UndoVariant", {
  enumerable: true,
  get: function () {
    return _UndoVariant.default;
  }
});
Object.defineProperty(exports, "UnfoldLessHorizontal", {
  enumerable: true,
  get: function () {
    return _UnfoldLessHorizontal.default;
  }
});
Object.defineProperty(exports, "UnfoldLessVertical", {
  enumerable: true,
  get: function () {
    return _UnfoldLessVertical.default;
  }
});
Object.defineProperty(exports, "UnfoldMoreHorizontal", {
  enumerable: true,
  get: function () {
    return _UnfoldMoreHorizontal.default;
  }
});
Object.defineProperty(exports, "UnfoldMoreVertical", {
  enumerable: true,
  get: function () {
    return _UnfoldMoreVertical.default;
  }
});
Object.defineProperty(exports, "Ungroup", {
  enumerable: true,
  get: function () {
    return _Ungroup.default;
  }
});
Object.defineProperty(exports, "Upload", {
  enumerable: true,
  get: function () {
    return _Upload.default;
  }
});
Object.defineProperty(exports, "VectorArrangeAbove", {
  enumerable: true,
  get: function () {
    return _VectorArrangeAbove.default;
  }
});
Object.defineProperty(exports, "VectorArrangeBelow", {
  enumerable: true,
  get: function () {
    return _VectorArrangeBelow.default;
  }
});
Object.defineProperty(exports, "VectorCombine", {
  enumerable: true,
  get: function () {
    return _VectorCombine.default;
  }
});
Object.defineProperty(exports, "VectorDifference", {
  enumerable: true,
  get: function () {
    return _VectorDifference.default;
  }
});
Object.defineProperty(exports, "VectorDifferenceAb", {
  enumerable: true,
  get: function () {
    return _VectorDifferenceAb.default;
  }
});
Object.defineProperty(exports, "VectorDifferenceBa", {
  enumerable: true,
  get: function () {
    return _VectorDifferenceBa.default;
  }
});
Object.defineProperty(exports, "VectorIntersection", {
  enumerable: true,
  get: function () {
    return _VectorIntersection.default;
  }
});
Object.defineProperty(exports, "VectorUnion", {
  enumerable: true,
  get: function () {
    return _VectorUnion.default;
  }
});
Object.defineProperty(exports, "ViewDashboard", {
  enumerable: true,
  get: function () {
    return _ViewDashboard.default;
  }
});
Object.defineProperty(exports, "ViewModule", {
  enumerable: true,
  get: function () {
    return _ViewModule.default;
  }
});
Object.defineProperty(exports, "VolumeHigh", {
  enumerable: true,
  get: function () {
    return _VolumeHigh.default;
  }
});
Object.defineProperty(exports, "VolumeMinus", {
  enumerable: true,
  get: function () {
    return _VolumeMinus.default;
  }
});
Object.defineProperty(exports, "VolumeMute", {
  enumerable: true,
  get: function () {
    return _VolumeMute.default;
  }
});
Object.defineProperty(exports, "VolumeOff", {
  enumerable: true,
  get: function () {
    return _VolumeOff.default;
  }
});
Object.defineProperty(exports, "VolumePlus", {
  enumerable: true,
  get: function () {
    return _VolumePlus.default;
  }
});
Object.defineProperty(exports, "Wifi", {
  enumerable: true,
  get: function () {
    return _Wifi.default;
  }
});
Object.defineProperty(exports, "Xml", {
  enumerable: true,
  get: function () {
    return _Xml.default;
  }
});

var _Account = _interopRequireDefault(require("./Account"));

var _Alarm = _interopRequireDefault(require("./Alarm"));

var _AlarmPlus = _interopRequireDefault(require("./AlarmPlus"));

var _Alert = _interopRequireDefault(require("./Alert"));

var _AlertCircle = _interopRequireDefault(require("./AlertCircle"));

var _AlertOctagon = _interopRequireDefault(require("./AlertOctagon"));

var _ArrangeBringForward = _interopRequireDefault(require("./ArrangeBringForward"));

var _ArrangeBringToFront = _interopRequireDefault(require("./ArrangeBringToFront"));

var _ArrangeSendBackward = _interopRequireDefault(require("./ArrangeSendBackward"));

var _ArrangeSendToBack = _interopRequireDefault(require("./ArrangeSendToBack"));

var _ArrowDown = _interopRequireDefault(require("./ArrowDown"));

var _ArrowDownCircle = _interopRequireDefault(require("./ArrowDownCircle"));

var _ArrowLeft = _interopRequireDefault(require("./ArrowLeft"));

var _ArrowLeftCircle = _interopRequireDefault(require("./ArrowLeftCircle"));

var _ArrowRight = _interopRequireDefault(require("./ArrowRight"));

var _ArrowRightCircle = _interopRequireDefault(require("./ArrowRightCircle"));

var _ArrowUp = _interopRequireDefault(require("./ArrowUp"));

var _ArrowUpCircle = _interopRequireDefault(require("./ArrowUpCircle"));

var _Bank = _interopRequireDefault(require("./Bank"));

var _Bell = _interopRequireDefault(require("./Bell"));

var _BellOff = _interopRequireDefault(require("./BellOff"));

var _Bluetooth = _interopRequireDefault(require("./Bluetooth"));

var _Book = _interopRequireDefault(require("./Book"));

var _BookMultiple = _interopRequireDefault(require("./BookMultiple"));

var _BookPlus = _interopRequireDefault(require("./BookPlus"));

var _Bookmark = _interopRequireDefault(require("./Bookmark"));

var _BorderAll = _interopRequireDefault(require("./BorderAll"));

var _BorderBottom = _interopRequireDefault(require("./BorderBottom"));

var _BorderHorizontal = _interopRequireDefault(require("./BorderHorizontal"));

var _BorderInside = _interopRequireDefault(require("./BorderInside"));

var _BorderLeft = _interopRequireDefault(require("./BorderLeft"));

var _BorderNone = _interopRequireDefault(require("./BorderNone"));

var _BorderOutside = _interopRequireDefault(require("./BorderOutside"));

var _BorderRight = _interopRequireDefault(require("./BorderRight"));

var _BorderTop = _interopRequireDefault(require("./BorderTop"));

var _BorderVertical = _interopRequireDefault(require("./BorderVertical"));

var _Briefcase = _interopRequireDefault(require("./Briefcase"));

var _Calendar = _interopRequireDefault(require("./Calendar"));

var _Camcorder = _interopRequireDefault(require("./Camcorder"));

var _Camera = _interopRequireDefault(require("./Camera"));

var _Cancel = _interopRequireDefault(require("./Cancel"));

var _Cart = _interopRequireDefault(require("./Cart"));

var _ChartAreaspline = _interopRequireDefault(require("./ChartAreaspline"));

var _ChartBar = _interopRequireDefault(require("./ChartBar"));

var _ChartHistogram = _interopRequireDefault(require("./ChartHistogram"));

var _ChartLine = _interopRequireDefault(require("./ChartLine"));

var _ChartPie = _interopRequireDefault(require("./ChartPie"));

var _Check = _interopRequireDefault(require("./Check"));

var _ChevronDoubleDown = _interopRequireDefault(require("./ChevronDoubleDown"));

var _ChevronDoubleLeft = _interopRequireDefault(require("./ChevronDoubleLeft"));

var _ChevronDoubleRight = _interopRequireDefault(require("./ChevronDoubleRight"));

var _ChevronDoubleUp = _interopRequireDefault(require("./ChevronDoubleUp"));

var _ChevronDown = _interopRequireDefault(require("./ChevronDown"));

var _ChevronLeft = _interopRequireDefault(require("./ChevronLeft"));

var _ChevronRight = _interopRequireDefault(require("./ChevronRight"));

var _ChevronUp = _interopRequireDefault(require("./ChevronUp"));

var _Circle = _interopRequireDefault(require("./Circle"));

var _Clipboard = _interopRequireDefault(require("./Clipboard"));

var _ClipboardCheck = _interopRequireDefault(require("./ClipboardCheck"));

var _ClipboardPlus = _interopRequireDefault(require("./ClipboardPlus"));

var _ClipboardText = _interopRequireDefault(require("./ClipboardText"));

var _Clock = _interopRequireDefault(require("./Clock"));

var _ClosedCaption = _interopRequireDefault(require("./ClosedCaption"));

var _Cloud = _interopRequireDefault(require("./Cloud"));

var _CloudDownload = _interopRequireDefault(require("./CloudDownload"));

var _CloudUpload = _interopRequireDefault(require("./CloudUpload"));

var _Comment = _interopRequireDefault(require("./Comment"));

var _CommentAlert = _interopRequireDefault(require("./CommentAlert"));

var _CommentText = _interopRequireDefault(require("./CommentText"));

var _Console = _interopRequireDefault(require("./Console"));

var _ContentCut = _interopRequireDefault(require("./ContentCut"));

var _ContentDuplicate = _interopRequireDefault(require("./ContentDuplicate"));

var _ContentPaste = _interopRequireDefault(require("./ContentPaste"));

var _ContentSave = _interopRequireDefault(require("./ContentSave"));

var _ContentSaveAll = _interopRequireDefault(require("./ContentSaveAll"));

var _CreditCard = _interopRequireDefault(require("./CreditCard"));

var _CreditCardScan = _interopRequireDefault(require("./CreditCardScan"));

var _Crop = _interopRequireDefault(require("./Crop"));

var _CropFree = _interopRequireDefault(require("./CropFree"));

var _CurrencyEur = _interopRequireDefault(require("./CurrencyEur"));

var _CurrencyGbp = _interopRequireDefault(require("./CurrencyGbp"));

var _CurrencyRub = _interopRequireDefault(require("./CurrencyRub"));

var _CurrencyUsd = _interopRequireDefault(require("./CurrencyUsd"));

var _Delete = _interopRequireDefault(require("./Delete"));

var _DiamondStone = _interopRequireDefault(require("./DiamondStone"));

var _DotsHorizontal = _interopRequireDefault(require("./DotsHorizontal"));

var _DotsVertical = _interopRequireDefault(require("./DotsVertical"));

var _Download = _interopRequireDefault(require("./Download"));

var _Eject = _interopRequireDefault(require("./Eject"));

var _Email = _interopRequireDefault(require("./Email"));

var _EmailOpen = _interopRequireDefault(require("./EmailOpen"));

var _Eye = _interopRequireDefault(require("./Eye"));

var _EyeOff = _interopRequireDefault(require("./EyeOff"));

var _Factory = _interopRequireDefault(require("./Factory"));

var _FastForward = _interopRequireDefault(require("./FastForward"));

var _File = _interopRequireDefault(require("./File"));

var _FileMultiple = _interopRequireDefault(require("./FileMultiple"));

var _FilePlus = _interopRequireDefault(require("./FilePlus"));

var _Flag = _interopRequireDefault(require("./Flag"));

var _Flash = _interopRequireDefault(require("./Flash"));

var _Flask = _interopRequireDefault(require("./Flask"));

var _FlaskEmpty = _interopRequireDefault(require("./FlaskEmpty"));

var _Folder = _interopRequireDefault(require("./Folder"));

var _FolderMultiple = _interopRequireDefault(require("./FolderMultiple"));

var _FolderPlus = _interopRequireDefault(require("./FolderPlus"));

var _FormatAlignBottom = _interopRequireDefault(require("./FormatAlignBottom"));

var _FormatAlignCenter = _interopRequireDefault(require("./FormatAlignCenter"));

var _FormatAlignJustify = _interopRequireDefault(require("./FormatAlignJustify"));

var _FormatAlignLeft = _interopRequireDefault(require("./FormatAlignLeft"));

var _FormatAlignMiddle = _interopRequireDefault(require("./FormatAlignMiddle"));

var _FormatAlignRight = _interopRequireDefault(require("./FormatAlignRight"));

var _FormatAlignTop = _interopRequireDefault(require("./FormatAlignTop"));

var _FormatBold = _interopRequireDefault(require("./FormatBold"));

var _FormatClear = _interopRequireDefault(require("./FormatClear"));

var _FormatFloatCenter = _interopRequireDefault(require("./FormatFloatCenter"));

var _FormatFloatLeft = _interopRequireDefault(require("./FormatFloatLeft"));

var _FormatFloatNone = _interopRequireDefault(require("./FormatFloatNone"));

var _FormatFloatRight = _interopRequireDefault(require("./FormatFloatRight"));

var _FormatIndentDecrease = _interopRequireDefault(require("./FormatIndentDecrease"));

var _FormatIndentIncrease = _interopRequireDefault(require("./FormatIndentIncrease"));

var _FormatItalic = _interopRequireDefault(require("./FormatItalic"));

var _FormatLineSpacing = _interopRequireDefault(require("./FormatLineSpacing"));

var _FormatListBulleted = _interopRequireDefault(require("./FormatListBulleted"));

var _FormatListChecks = _interopRequireDefault(require("./FormatListChecks"));

var _FormatListNumbers = _interopRequireDefault(require("./FormatListNumbers"));

var _FormatQuoteClose = _interopRequireDefault(require("./FormatQuoteClose"));

var _FormatQuoteOpen = _interopRequireDefault(require("./FormatQuoteOpen"));

var _FormatUnderline = _interopRequireDefault(require("./FormatUnderline"));

var _FormatWrapInline = _interopRequireDefault(require("./FormatWrapInline"));

var _FormatWrapSquare = _interopRequireDefault(require("./FormatWrapSquare"));

var _FormatWrapTight = _interopRequireDefault(require("./FormatWrapTight"));

var _FormatWrapTopBottom = _interopRequireDefault(require("./FormatWrapTopBottom"));

var _Forum = _interopRequireDefault(require("./Forum"));

var _Fullscreen = _interopRequireDefault(require("./Fullscreen"));

var _FullscreenExit = _interopRequireDefault(require("./FullscreenExit"));

var _Gift = _interopRequireDefault(require("./Gift"));

var _Grid = _interopRequireDefault(require("./Grid"));

var _GridLarge = _interopRequireDefault(require("./GridLarge"));

var _GridOff = _interopRequireDefault(require("./GridOff"));

var _Group = _interopRequireDefault(require("./Group"));

var _Hamburger = _interopRequireDefault(require("./Hamburger"));

var _Heart = _interopRequireDefault(require("./Heart"));

var _HeartHalf = _interopRequireDefault(require("./HeartHalf"));

var _HeartOff = _interopRequireDefault(require("./HeartOff"));

var _Help = _interopRequireDefault(require("./Help"));

var _HelpCircle = _interopRequireDefault(require("./HelpCircle"));

var _Hexagon = _interopRequireDefault(require("./Hexagon"));

var _Home = _interopRequireDefault(require("./Home"));

var _Image = _interopRequireDefault(require("./Image"));

var _Inbox = _interopRequireDefault(require("./Inbox"));

var _Information = _interopRequireDefault(require("./Information"));

var _Label = _interopRequireDefault(require("./Label"));

var _Layers = _interopRequireDefault(require("./Layers"));

var _Lightbulb = _interopRequireDefault(require("./Lightbulb"));

var _LightbulbOn = _interopRequireDefault(require("./LightbulbOn"));

var _Link = _interopRequireDefault(require("./Link"));

var _LinkVariant = _interopRequireDefault(require("./LinkVariant"));

var _Lock = _interopRequireDefault(require("./Lock"));

var _LockOpen = _interopRequireDefault(require("./LockOpen"));

var _Login = _interopRequireDefault(require("./Login"));

var _Logout = _interopRequireDefault(require("./Logout"));

var _Magnify = _interopRequireDefault(require("./Magnify"));

var _MagnifyMinus = _interopRequireDefault(require("./MagnifyMinus"));

var _MagnifyPlus = _interopRequireDefault(require("./MagnifyPlus"));

var _MapMarker = _interopRequireDefault(require("./MapMarker"));

var _Memory = _interopRequireDefault(require("./Memory"));

var _Menu = _interopRequireDefault(require("./Menu"));

var _Message = _interopRequireDefault(require("./Message"));

var _MessageAlert = _interopRequireDefault(require("./MessageAlert"));

var _MessageImage = _interopRequireDefault(require("./MessageImage"));

var _MessageProcessing = _interopRequireDefault(require("./MessageProcessing"));

var _MessageReply = _interopRequireDefault(require("./MessageReply"));

var _MessageText = _interopRequireDefault(require("./MessageText"));

var _MessageVideo = _interopRequireDefault(require("./MessageVideo"));

var _Microphone = _interopRequireDefault(require("./Microphone"));

var _MicrophoneOff = _interopRequireDefault(require("./MicrophoneOff"));

var _Minus = _interopRequireDefault(require("./Minus"));

var _MinusBox = _interopRequireDefault(require("./MinusBox"));

var _MinusCircle = _interopRequireDefault(require("./MinusCircle"));

var _Monitor = _interopRequireDefault(require("./Monitor"));

var _MonitorMultiple = _interopRequireDefault(require("./MonitorMultiple"));

var _Music = _interopRequireDefault(require("./Music"));

var _MusicOff = _interopRequireDefault(require("./MusicOff"));

var _NfcVariant = _interopRequireDefault(require("./NfcVariant"));

var _Note = _interopRequireDefault(require("./Note"));

var _NoteMultiple = _interopRequireDefault(require("./NoteMultiple"));

var _NotePlus = _interopRequireDefault(require("./NotePlus"));

var _NoteText = _interopRequireDefault(require("./NoteText"));

var _Octagon = _interopRequireDefault(require("./Octagon"));

var _Paperclip = _interopRequireDefault(require("./Paperclip"));

var _Pause = _interopRequireDefault(require("./Pause"));

var _Pencil = _interopRequireDefault(require("./Pencil"));

var _Phone = _interopRequireDefault(require("./Phone"));

var _Pin = _interopRequireDefault(require("./Pin"));

var _PinOff = _interopRequireDefault(require("./PinOff"));

var _Play = _interopRequireDefault(require("./Play"));

var _Plus = _interopRequireDefault(require("./Plus"));

var _PlusBox = _interopRequireDefault(require("./PlusBox"));

var _PlusCircle = _interopRequireDefault(require("./PlusCircle"));

var _Power = _interopRequireDefault(require("./Power"));

var _Presentation = _interopRequireDefault(require("./Presentation"));

var _PresentationPlay = _interopRequireDefault(require("./PresentationPlay"));

var _Printer = _interopRequireDefault(require("./Printer"));

var _RedoVariant = _interopRequireDefault(require("./RedoVariant"));

var _Refresh = _interopRequireDefault(require("./Refresh"));

var _Repeat = _interopRequireDefault(require("./Repeat"));

var _RepeatOff = _interopRequireDefault(require("./RepeatOff"));

var _RepeatOnce = _interopRequireDefault(require("./RepeatOnce"));

var _Rewind = _interopRequireDefault(require("./Rewind"));

var _Rhombus = _interopRequireDefault(require("./Rhombus"));

var _Rss = _interopRequireDefault(require("./Rss"));

var _Script = _interopRequireDefault(require("./Script"));

var _Settings = _interopRequireDefault(require("./Settings"));

var _Shield = _interopRequireDefault(require("./Shield"));

var _Shuffle = _interopRequireDefault(require("./Shuffle"));

var _Signal = _interopRequireDefault(require("./Signal"));

var _Sitemap = _interopRequireDefault(require("./Sitemap"));

var _SkipNext = _interopRequireDefault(require("./SkipNext"));

var _SkipPrevious = _interopRequireDefault(require("./SkipPrevious"));

var _Sleep = _interopRequireDefault(require("./Sleep"));

var _SleepOff = _interopRequireDefault(require("./SleepOff"));

var _Spellcheck = _interopRequireDefault(require("./Spellcheck"));

var _Square = _interopRequireDefault(require("./Square"));

var _Star = _interopRequireDefault(require("./Star"));

var _StarHalf = _interopRequireDefault(require("./StarHalf"));

var _Stop = _interopRequireDefault(require("./Stop"));

var _Tab = _interopRequireDefault(require("./Tab"));

var _TabPlus = _interopRequireDefault(require("./TabPlus"));

var _Table = _interopRequireDefault(require("./Table"));

var _Taco = _interopRequireDefault(require("./Taco"));

var _Tag = _interopRequireDefault(require("./Tag"));

var _Television = _interopRequireDefault(require("./Television"));

var _ThumbDown = _interopRequireDefault(require("./ThumbDown"));

var _ThumbUp = _interopRequireDefault(require("./ThumbUp"));

var _ThumbsUpDown = _interopRequireDefault(require("./ThumbsUpDown"));

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

var _TooltipText = _interopRequireDefault(require("./TooltipText"));

var _Triangle = _interopRequireDefault(require("./Triangle"));

var _Trophy = _interopRequireDefault(require("./Trophy"));

var _UndoVariant = _interopRequireDefault(require("./UndoVariant"));

var _UnfoldLessHorizontal = _interopRequireDefault(require("./UnfoldLessHorizontal"));

var _UnfoldLessVertical = _interopRequireDefault(require("./UnfoldLessVertical"));

var _UnfoldMoreHorizontal = _interopRequireDefault(require("./UnfoldMoreHorizontal"));

var _UnfoldMoreVertical = _interopRequireDefault(require("./UnfoldMoreVertical"));

var _Ungroup = _interopRequireDefault(require("./Ungroup"));

var _Upload = _interopRequireDefault(require("./Upload"));

var _VectorArrangeAbove = _interopRequireDefault(require("./VectorArrangeAbove"));

var _VectorArrangeBelow = _interopRequireDefault(require("./VectorArrangeBelow"));

var _VectorCombine = _interopRequireDefault(require("./VectorCombine"));

var _VectorDifference = _interopRequireDefault(require("./VectorDifference"));

var _VectorDifferenceAb = _interopRequireDefault(require("./VectorDifferenceAb"));

var _VectorDifferenceBa = _interopRequireDefault(require("./VectorDifferenceBa"));

var _VectorIntersection = _interopRequireDefault(require("./VectorIntersection"));

var _VectorUnion = _interopRequireDefault(require("./VectorUnion"));

var _ViewDashboard = _interopRequireDefault(require("./ViewDashboard"));

var _ViewModule = _interopRequireDefault(require("./ViewModule"));

var _VolumeHigh = _interopRequireDefault(require("./VolumeHigh"));

var _VolumeMinus = _interopRequireDefault(require("./VolumeMinus"));

var _VolumeMute = _interopRequireDefault(require("./VolumeMute"));

var _VolumeOff = _interopRequireDefault(require("./VolumeOff"));

var _VolumePlus = _interopRequireDefault(require("./VolumePlus"));

var _Wifi = _interopRequireDefault(require("./Wifi"));

var _Xml = _interopRequireDefault(require("./Xml"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
