// devextreme-react/index.js
/*!
 * devextreme-react
 * Version: 20.2.6
 * Build date: Tue Mar 16 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-react
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var template_1 = require("./core/template");
Object.defineProperty(exports, "Template", { enumerable: true, get: function () { return template_1.Template; } });
var accordion_1 = require("./accordion");
Object.defineProperty(exports, "Accordion", { enumerable: true, get: function () { return accordion_1.Accordion; } });
var action_sheet_1 = require("./action-sheet");
Object.defineProperty(exports, "ActionSheet", { enumerable: true, get: function () { return action_sheet_1.ActionSheet; } });
var autocomplete_1 = require("./autocomplete");
Object.defineProperty(exports, "Autocomplete", { enumerable: true, get: function () { return autocomplete_1.Autocomplete; } });
var bar_gauge_1 = require("./bar-gauge");
Object.defineProperty(exports, "BarGauge", { enumerable: true, get: function () { return bar_gauge_1.BarGauge; } });
var box_1 = require("./box");
Object.defineProperty(exports, "Box", { enumerable: true, get: function () { return box_1.Box; } });
var bullet_1 = require("./bullet");
Object.defineProperty(exports, "Bullet", { enumerable: true, get: function () { return bullet_1.Bullet; } });
var button_1 = require("./button");
Object.defineProperty(exports, "Button", { enumerable: true, get: function () { return button_1.Button; } });
var button_group_1 = require("./button-group");
Object.defineProperty(exports, "ButtonGroup", { enumerable: true, get: function () { return button_group_1.ButtonGroup; } });
var calendar_1 = require("./calendar");
Object.defineProperty(exports, "Calendar", { enumerable: true, get: function () { return calendar_1.Calendar; } });
var chart_1 = require("./chart");
Object.defineProperty(exports, "Chart", { enumerable: true, get: function () { return chart_1.Chart; } });
var check_box_1 = require("./check-box");
Object.defineProperty(exports, "CheckBox", { enumerable: true, get: function () { return check_box_1.CheckBox; } });
var circular_gauge_1 = require("./circular-gauge");
Object.defineProperty(exports, "CircularGauge", { enumerable: true, get: function () { return circular_gauge_1.CircularGauge; } });
var color_box_1 = require("./color-box");
Object.defineProperty(exports, "ColorBox", { enumerable: true, get: function () { return color_box_1.ColorBox; } });
var context_menu_1 = require("./context-menu");
Object.defineProperty(exports, "ContextMenu", { enumerable: true, get: function () { return context_menu_1.ContextMenu; } });
var data_grid_1 = require("./data-grid");
Object.defineProperty(exports, "DataGrid", { enumerable: true, get: function () { return data_grid_1.DataGrid; } });
var date_box_1 = require("./date-box");
Object.defineProperty(exports, "DateBox", { enumerable: true, get: function () { return date_box_1.DateBox; } });
var defer_rendering_1 = require("./defer-rendering");
Object.defineProperty(exports, "DeferRendering", { enumerable: true, get: function () { return defer_rendering_1.DeferRendering; } });
var diagram_1 = require("./diagram");
Object.defineProperty(exports, "Diagram", { enumerable: true, get: function () { return diagram_1.Diagram; } });
var draggable_1 = require("./draggable");
Object.defineProperty(exports, "Draggable", { enumerable: true, get: function () { return draggable_1.Draggable; } });
var drawer_1 = require("./drawer");
Object.defineProperty(exports, "Drawer", { enumerable: true, get: function () { return drawer_1.Drawer; } });
var drop_down_box_1 = require("./drop-down-box");
Object.defineProperty(exports, "DropDownBox", { enumerable: true, get: function () { return drop_down_box_1.DropDownBox; } });
var drop_down_button_1 = require("./drop-down-button");
Object.defineProperty(exports, "DropDownButton", { enumerable: true, get: function () { return drop_down_button_1.DropDownButton; } });
var file_manager_1 = require("./file-manager");
Object.defineProperty(exports, "FileManager", { enumerable: true, get: function () { return file_manager_1.FileManager; } });
var file_uploader_1 = require("./file-uploader");
Object.defineProperty(exports, "FileUploader", { enumerable: true, get: function () { return file_uploader_1.FileUploader; } });
var filter_builder_1 = require("./filter-builder");
Object.defineProperty(exports, "FilterBuilder", { enumerable: true, get: function () { return filter_builder_1.FilterBuilder; } });
var form_1 = require("./form");
Object.defineProperty(exports, "Form", { enumerable: true, get: function () { return form_1.Form; } });
var funnel_1 = require("./funnel");
Object.defineProperty(exports, "Funnel", { enumerable: true, get: function () { return funnel_1.Funnel; } });
var gallery_1 = require("./gallery");
Object.defineProperty(exports, "Gallery", { enumerable: true, get: function () { return gallery_1.Gallery; } });
var gantt_1 = require("./gantt");
Object.defineProperty(exports, "Gantt", { enumerable: true, get: function () { return gantt_1.Gantt; } });
var html_editor_1 = require("./html-editor");
Object.defineProperty(exports, "HtmlEditor", { enumerable: true, get: function () { return html_editor_1.HtmlEditor; } });
var linear_gauge_1 = require("./linear-gauge");
Object.defineProperty(exports, "LinearGauge", { enumerable: true, get: function () { return linear_gauge_1.LinearGauge; } });
var list_1 = require("./list");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return list_1.List; } });
var load_indicator_1 = require("./load-indicator");
Object.defineProperty(exports, "LoadIndicator", { enumerable: true, get: function () { return load_indicator_1.LoadIndicator; } });
var load_panel_1 = require("./load-panel");
Object.defineProperty(exports, "LoadPanel", { enumerable: true, get: function () { return load_panel_1.LoadPanel; } });
var lookup_1 = require("./lookup");
Object.defineProperty(exports, "Lookup", { enumerable: true, get: function () { return lookup_1.Lookup; } });
var map_1 = require("./map");
Object.defineProperty(exports, "Map", { enumerable: true, get: function () { return map_1.Map; } });
var menu_1 = require("./menu");
Object.defineProperty(exports, "Menu", { enumerable: true, get: function () { return menu_1.Menu; } });
var multi_view_1 = require("./multi-view");
Object.defineProperty(exports, "MultiView", { enumerable: true, get: function () { return multi_view_1.MultiView; } });
var nav_bar_1 = require("./nav-bar");
Object.defineProperty(exports, "NavBar", { enumerable: true, get: function () { return nav_bar_1.NavBar; } });
var number_box_1 = require("./number-box");
Object.defineProperty(exports, "NumberBox", { enumerable: true, get: function () { return number_box_1.NumberBox; } });
var pie_chart_1 = require("./pie-chart");
Object.defineProperty(exports, "PieChart", { enumerable: true, get: function () { return pie_chart_1.PieChart; } });
var pivot_grid_1 = require("./pivot-grid");
Object.defineProperty(exports, "PivotGrid", { enumerable: true, get: function () { return pivot_grid_1.PivotGrid; } });
var pivot_grid_field_chooser_1 = require("./pivot-grid-field-chooser");
Object.defineProperty(exports, "PivotGridFieldChooser", { enumerable: true, get: function () { return pivot_grid_field_chooser_1.PivotGridFieldChooser; } });
var polar_chart_1 = require("./polar-chart");
Object.defineProperty(exports, "PolarChart", { enumerable: true, get: function () { return polar_chart_1.PolarChart; } });
var popover_1 = require("./popover");
Object.defineProperty(exports, "Popover", { enumerable: true, get: function () { return popover_1.Popover; } });
var popup_1 = require("./popup");
Object.defineProperty(exports, "Popup", { enumerable: true, get: function () { return popup_1.Popup; } });
var progress_bar_1 = require("./progress-bar");
Object.defineProperty(exports, "ProgressBar", { enumerable: true, get: function () { return progress_bar_1.ProgressBar; } });
var radio_group_1 = require("./radio-group");
Object.defineProperty(exports, "RadioGroup", { enumerable: true, get: function () { return radio_group_1.RadioGroup; } });
var range_selector_1 = require("./range-selector");
Object.defineProperty(exports, "RangeSelector", { enumerable: true, get: function () { return range_selector_1.RangeSelector; } });
var range_slider_1 = require("./range-slider");
Object.defineProperty(exports, "RangeSlider", { enumerable: true, get: function () { return range_slider_1.RangeSlider; } });
var recurrence_editor_1 = require("./recurrence-editor");
Object.defineProperty(exports, "RecurrenceEditor", { enumerable: true, get: function () { return recurrence_editor_1.RecurrenceEditor; } });
var resizable_1 = require("./resizable");
Object.defineProperty(exports, "Resizable", { enumerable: true, get: function () { return resizable_1.Resizable; } });
var responsive_box_1 = require("./responsive-box");
Object.defineProperty(exports, "ResponsiveBox", { enumerable: true, get: function () { return responsive_box_1.ResponsiveBox; } });
var sankey_1 = require("./sankey");
Object.defineProperty(exports, "Sankey", { enumerable: true, get: function () { return sankey_1.Sankey; } });
var scheduler_1 = require("./scheduler");
Object.defineProperty(exports, "Scheduler", { enumerable: true, get: function () { return scheduler_1.Scheduler; } });
var scroll_view_1 = require("./scroll-view");
Object.defineProperty(exports, "ScrollView", { enumerable: true, get: function () { return scroll_view_1.ScrollView; } });
var select_box_1 = require("./select-box");
Object.defineProperty(exports, "SelectBox", { enumerable: true, get: function () { return select_box_1.SelectBox; } });
var slide_out_1 = require("./slide-out");
Object.defineProperty(exports, "SlideOut", { enumerable: true, get: function () { return slide_out_1.SlideOut; } });
var slide_out_view_1 = require("./slide-out-view");
Object.defineProperty(exports, "SlideOutView", { enumerable: true, get: function () { return slide_out_view_1.SlideOutView; } });
var slider_1 = require("./slider");
Object.defineProperty(exports, "Slider", { enumerable: true, get: function () { return slider_1.Slider; } });
var sortable_1 = require("./sortable");
Object.defineProperty(exports, "Sortable", { enumerable: true, get: function () { return sortable_1.Sortable; } });
var sparkline_1 = require("./sparkline");
Object.defineProperty(exports, "Sparkline", { enumerable: true, get: function () { return sparkline_1.Sparkline; } });
var speed_dial_action_1 = require("./speed-dial-action");
Object.defineProperty(exports, "SpeedDialAction", { enumerable: true, get: function () { return speed_dial_action_1.SpeedDialAction; } });
var switch_1 = require("./switch");
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return switch_1.Switch; } });
var tab_panel_1 = require("./tab-panel");
Object.defineProperty(exports, "TabPanel", { enumerable: true, get: function () { return tab_panel_1.TabPanel; } });
var tabs_1 = require("./tabs");
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return tabs_1.Tabs; } });
var tag_box_1 = require("./tag-box");
Object.defineProperty(exports, "TagBox", { enumerable: true, get: function () { return tag_box_1.TagBox; } });
var text_area_1 = require("./text-area");
Object.defineProperty(exports, "TextArea", { enumerable: true, get: function () { return text_area_1.TextArea; } });
var text_box_1 = require("./text-box");
Object.defineProperty(exports, "TextBox", { enumerable: true, get: function () { return text_box_1.TextBox; } });
var tile_view_1 = require("./tile-view");
Object.defineProperty(exports, "TileView", { enumerable: true, get: function () { return tile_view_1.TileView; } });
var toast_1 = require("./toast");
Object.defineProperty(exports, "Toast", { enumerable: true, get: function () { return toast_1.Toast; } });
var toolbar_1 = require("./toolbar");
Object.defineProperty(exports, "Toolbar", { enumerable: true, get: function () { return toolbar_1.Toolbar; } });
var tooltip_1 = require("./tooltip");
Object.defineProperty(exports, "Tooltip", { enumerable: true, get: function () { return tooltip_1.Tooltip; } });
var tree_list_1 = require("./tree-list");
Object.defineProperty(exports, "TreeList", { enumerable: true, get: function () { return tree_list_1.TreeList; } });
var tree_map_1 = require("./tree-map");
Object.defineProperty(exports, "TreeMap", { enumerable: true, get: function () { return tree_map_1.TreeMap; } });
var tree_view_1 = require("./tree-view");
Object.defineProperty(exports, "TreeView", { enumerable: true, get: function () { return tree_view_1.TreeView; } });
var validation_group_1 = require("./validation-group");
Object.defineProperty(exports, "ValidationGroup", { enumerable: true, get: function () { return validation_group_1.ValidationGroup; } });
var validation_summary_1 = require("./validation-summary");
Object.defineProperty(exports, "ValidationSummary", { enumerable: true, get: function () { return validation_summary_1.ValidationSummary; } });
var validator_1 = require("./validator");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validator_1.Validator; } });
var vector_map_1 = require("./vector-map");
Object.defineProperty(exports, "VectorMap", { enumerable: true, get: function () { return vector_map_1.VectorMap; } });
