// @vime/core/dist/cjs/ProviderConnect-3c00ee3c.js
'use strict';

const index = require('./index-9e4d61ac.js');
const PlayerProps = require('./PlayerProps-4bbfc16a.js');
const withComponentRegistry = require('./withComponentRegistry-50d191fe.js');
const Provider = require('./Provider-0e71f5a4.js');

const PROVIDER_CHANGE_EVENT = 'vmProviderChange';
/**
 * Creates a dispatcher on the given `ref`, to send updates to the closest ancestor player via
 * the custom `vmProviderChange` event.
 *
 * @param ref A component reference to dispatch the state change events from.
 */
const createProviderDispatcher = (ref) => (prop, value) => {
  const el = withComponentRegistry.isInstanceOf(ref, HTMLElement) ? ref : index.getElement(ref);
  const event = new CustomEvent(PROVIDER_CHANGE_EVENT, {
    bubbles: true,
    composed: true,
    detail: { by: el, prop, value },
  });
  el.dispatchEvent(event);
};

const providerWritableProps = new Set([
  'ready',
  'playing',
  'playbackReady',
  'playbackStarted',
  'playbackEnded',
  'seeking',
  'buffered',
  'buffering',
  'duration',
  'viewType',
  'mediaTitle',
  'mediaType',
  'currentSrc',
  'currentPoster',
  'playbackRates',
  'playbackQualities',
  'textTracks',
  'currentTextTrack',
  'isTextTrackVisible',
  'audioTracks',
  'currentAudioTrack',
  'isPiPActive',
  'isFullscreenActive',
]);
const isProviderWritableProp = (prop) => PlayerProps.isWritableProp(prop) || providerWritableProps.has(prop);

const PROVIDER_CACHE_KEY = Symbol('vmProviderCache');
const PROVIDER_CONNECT_EVENT = 'vmMediaProviderConnect';
const PROVIDER_DISCONNECT_EVENT = 'vmMediaProviderDisconnect';
function buildProviderConnectEvent(name, host) {
  return new CustomEvent(name, {
    bubbles: true,
    composed: true,
    detail: host,
  });
}
function withProviderHost(connector) {
  const el = index.getElement(connector);
  const disposal = new withComponentRegistry.Disposal();
  const cache = new Map();
  connector[PROVIDER_CACHE_KEY] = cache;
  function initCache() {
    Object.keys(connector).forEach((prop) => {
      cache.set(prop, connector[prop]);
    });
  }
  function onDisconnect() {
    index.writeTask(async () => {
      connector.ready = false;
      connector.provider = undefined;
      cache.clear();
      connector.onProviderDisconnect?.();
      el.dispatchEvent(buildProviderConnectEvent(PROVIDER_DISCONNECT_EVENT));
    });
  }
  function onConnect(event) {
    event.stopImmediatePropagation();
    initCache();
    const hostRef = event.detail;
    const host = index.getElement(event.detail);
    if (connector.provider === host)
      return;
    const name = host?.nodeName.toLowerCase().replace('vm-', '');
    index.writeTask(async () => {
      connector.provider = host;
      connector.currentProvider = Object.values(Provider.Provider).find((provider) => name === provider);
      withComponentRegistry.createStencilHook(hostRef, undefined, () => onDisconnect());
    });
  }
  function onChange(event) {
    event.stopImmediatePropagation();
    const { by, prop, value } = event.detail;
    if (!isProviderWritableProp(prop)) {
      connector.logger?.warn(`${by.nodeName} tried to change \`${prop}\` but it is readonly.`);
      return;
    }
    index.writeTask(() => {
      cache.set(prop, value);
      connector[prop] = value;
    });
  }
  withComponentRegistry.createStencilHook(connector, () => {
    disposal.add(withComponentRegistry.listen(el, PROVIDER_CONNECT_EVENT, onConnect));
    disposal.add(withComponentRegistry.listen(el, PROVIDER_CHANGE_EVENT, onChange));
  }, () => {
    disposal.empty();
    cache.clear();
  });
}
function withProviderConnect(ref) {
  const connectEvent = buildProviderConnectEvent(PROVIDER_CONNECT_EVENT, ref);
  withComponentRegistry.createStencilHook(ref, () => {
    index.getElement(ref).dispatchEvent(connectEvent);
  });
}

exports.PROVIDER_CACHE_KEY = PROVIDER_CACHE_KEY;
exports.createProviderDispatcher = createProviderDispatcher;
exports.withProviderConnect = withProviderConnect;
exports.withProviderHost = withProviderHost;
