// @vime/core/dist/cjs/vm-audio.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9e4d61ac.js');
require('./PlayerProps-4bbfc16a.js');
const withComponentRegistry = require('./withComponentRegistry-50d191fe.js');
const ViewType = require('./ViewType-435b01ff.js');
require('./Provider-0e71f5a4.js');
const utils = require('./utils-b8b7354f.js');
const ProviderConnect = require('./ProviderConnect-3c00ee3c.js');

const Audio = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * @internal Whether an external SDK will attach itself to the media player and control it.
     */
    this.willAttach = false;
    /** @inheritdoc */
    this.preload = 'metadata';
    withComponentRegistry.withComponentRegistry(this);
    if (!this.willAttach)
      ProviderConnect.withProviderConnect(this);
  }
  /** @internal */
  async getAdapter() {
    const adapter = (await this.fileProvider?.getAdapter()) ?? {};
    adapter.canPlay = async (type) => withComponentRegistry.isString(type) && utils.audioRegex.test(type);
    return adapter;
  }
  render() {
    return (
    // @ts-ignore
    index.h("vm-file", { noConnect: true, willAttach: this.willAttach, crossOrigin: this.crossOrigin, preload: this.preload, disableRemotePlayback: this.disableRemotePlayback, mediaTitle: this.mediaTitle, viewType: ViewType.ViewType.Audio, ref: (el) => {
        this.fileProvider = el;
      } }, index.h("slot", null)));
  }
};

exports.vm_audio = Audio;
