// @vime/core/dist/cjs/vm-default-ui.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9e4d61ac.js');
const withComponentRegistry = require('./withComponentRegistry-50d191fe.js');

const defaultUiCss = ":host{display:contents;pointer-events:none}";

const DefaultUI = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Whether clicking the player should not toggle playback.
     */
    this.noClickToPlay = false;
    /**
     * Whether double clicking the player should not toggle fullscreen mode.
     */
    this.noDblClickFullscreen = false;
    /**
     * Whether the custom captions UI should not be loaded.
     */
    this.noCaptions = false;
    /**
     * Whether the custom poster UI should not be loaded.
     */
    this.noPoster = false;
    /**
     * Whether the custom spinner UI should not be loaded.
     */
    this.noSpinner = false;
    /**
     * Whether the custom default controls should not be loaded.
     */
    this.noControls = false;
    /**
     * Whether the custom default settings menu should not be loaded.
     */
    this.noSettings = false;
    /**
     * Whether the default loading screen should not be loaded.
     */
    this.noLoadingScreen = false;
    withComponentRegistry.withComponentRegistry(this);
  }
  render() {
    return (index.h("vm-ui", null, !this.noClickToPlay && index.h("vm-click-to-play", null), !this.noDblClickFullscreen && index.h("vm-dbl-click-fullscreen", null), !this.noCaptions && index.h("vm-captions", null), !this.noPoster && index.h("vm-poster", null), !this.noSpinner && index.h("vm-spinner", null), !this.noLoadingScreen && index.h("vm-loading-screen", null), !this.noControls && index.h("vm-default-controls", null), !this.noSettings && index.h("vm-default-settings", null), index.h("slot", null)));
  }
};
DefaultUI.style = defaultUiCss;

exports.vm_default_ui = DefaultUI;
