// @vime/core/dist/collection/components/ui/settings/menu-radio-group/menu-radio-group.js
import { h, Component, Element, Prop, Watch, Event, Listen, } from '@stencil/core';
import { withComponentRegistry } from '../../../core/player/withComponentRegistry';
/**
 * @slot - Used to pass in radio buttons (`vm-menu-radio`).
 */
export class MenuRadioGroup {
  constructor() {
    withComponentRegistry(this);
  }
  onValueChange() {
    this.findRadios()?.forEach((radio) => {
      radio.checked = radio.value === this.value;
    });
  }
  connectedCallback() {
    this.onValueChange();
  }
  componentDidLoad() {
    this.onValueChange();
  }
  onSelectionChange(event) {
    const radio = event.target;
    this.value = radio.value;
  }
  findRadios() {
    return this.host
      .shadowRoot.querySelector('slot')
      ?.assignedElements();
  }
  render() {
    return h("slot", null);
  }
  static get is() { return "vm-menu-radio-group"; }
  static get encapsulation() { return "shadow"; }
  static get properties() { return {
    "value": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "The current value selected for this group."
      },
      "attribute": "value",
      "reflect": false
    }
  }; }
  static get events() { return [{
      "method": "vmCheck",
      "name": "vmCheck",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Emitted when a new radio button is selected for this group."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }]; }
  static get elementRef() { return "host"; }
  static get watchers() { return [{
      "propName": "value",
      "methodName": "onValueChange"
    }]; }
  static get listeners() { return [{
      "name": "vmCheck",
      "method": "onSelectionChange",
      "target": undefined,
      "capture": false,
      "passive": false
    }]; }
}
