// @vime/core/dist/collection/components/ui/settings/default-settings/default-settings.js
import { h, Component, Prop, State, Watch } from '@stencil/core';
import { withPlayerContext } from '../../../core/player/withPlayerContext';
import { Disposal } from '../../../../utils/Disposal';
import { createDispatcher, } from '../../../core/player/PlayerDispatcher';
import { findPlayer } from '../../../core/player/findPlayer';
import { getPlayerFromRegistry, withComponentRegistry, } from '../../../core/player/withComponentRegistry';
/**
 * @slot - Used to extend the settings with additional menu options (see `vm-submenu` or
 * `vm-menu-item`).
 */
export class DefaultSettings {
  constructor() {
    this.textTracksDisposal = new Disposal();
    this.canSetPlaybackRate = false;
    this.canSetPlaybackQuality = false;
    this.canSetTextTrack = false;
    this.canSetAudioTrack = false;
    /**
     * Pins the settings to the defined position inside the video player. This has no effect when
     * the view is of type `audio`, it will always be `bottomRight`.
     */
    this.pin = 'bottomRight';
    /** @internal */
    this.i18n = {};
    /** @internal */
    this.playbackReady = false;
    /** @internal */
    this.playbackRate = 1;
    /** @internal */
    this.playbackRates = [1];
    /** @internal */
    this.isVideoView = false;
    /** @internal */
    this.playbackQualities = [];
    /** @internal */
    this.textTracks = [];
    /** @internal */
    this.currentTextTrack = -1;
    /** @internal */
    this.audioTracks = [];
    /** @internal */
    this.currentAudioTrack = -1;
    /** @internal */
    this.isTextTrackVisible = true;
    withComponentRegistry(this);
    withPlayerContext(this, [
      'i18n',
      'playbackReady',
      'playbackRate',
      'playbackRates',
      'playbackQuality',
      'playbackQualities',
      'isVideoView',
      'textTracks',
      'currentTextTrack',
      'isTextTrackVisible',
      'audioTracks',
      'currentAudioTrack',
    ]);
  }
  async onPlaybackReady() {
    const player = await findPlayer(this);
    this.canSetPlaybackQuality = await player.canSetPlaybackQuality();
    this.canSetPlaybackRate = await player.canSetPlaybackRate();
  }
  async onAudioTracksChange() {
    const player = getPlayerFromRegistry(this);
    this.canSetAudioTrack = (await player?.canSetAudioTrack()) ?? false;
  }
  async onTextTracksChange() {
    const player = getPlayerFromRegistry(this);
    this.canSetTextTrack = (await player?.canSetTextTrack()) ?? false;
  }
  connectedCallback() {
    this.dispatch = createDispatcher(this);
  }
  componentDidLoad() {
    this.onTextTracksChange();
  }
  disconnectedCallback() {
    this.textTracksDisposal.empty();
  }
  onPlaybackRateSelect(event) {
    const radio = event.target;
    this.dispatch('playbackRate', parseFloat(radio.value));
  }
  buildPlaybackRateSubmenu() {
    if (this.playbackRates.length <= 1 || !this.canSetPlaybackRate) {
      return (h("vm-menu-item", { label: this.i18n.playbackRate, hint: this.i18n.normal }));
    }
    const formatRate = (rate) => rate === 1 ? this.i18n.normal : `${rate}`;
    return (h("vm-submenu", { label: this.i18n.playbackRate, hint: formatRate(this.playbackRate) },
      h("vm-menu-radio-group", { value: `${this.playbackRate}`, onVmCheck: this.onPlaybackRateSelect.bind(this) }, this.playbackRates.map((rate) => (h("vm-menu-radio", { label: formatRate(rate), value: `${rate}` }))))));
  }
  onPlaybackQualitySelect(event) {
    const radio = event.target;
    this.dispatch('playbackQuality', radio.value);
  }
  buildPlaybackQualitySubmenu() {
    if (this.playbackQualities.length <= 1 || !this.canSetPlaybackQuality) {
      return (h("vm-menu-item", { label: this.i18n.playbackQuality, hint: this.playbackQuality ?? this.i18n.auto }));
    }
    // @TODO this doesn't account for audio qualities yet.
    const getBadge = (quality) => {
      const verticalPixels = parseInt(quality.slice(0, -1), 10);
      if (verticalPixels >= 2160)
        return 'UHD';
      if (verticalPixels >= 1080)
        return 'HD';
      return undefined;
    };
    return (h("vm-submenu", { label: this.i18n.playbackQuality, hint: this.playbackQuality },
      h("vm-menu-radio-group", { value: this.playbackQuality, onVmCheck: this.onPlaybackQualitySelect.bind(this) }, this.playbackQualities.map((quality) => (h("vm-menu-radio", { label: quality, value: quality, badge: getBadge(quality) }))))));
  }
  onTextTrackSelect(event) {
    const radio = event.target;
    const trackId = parseInt(radio.value, 10);
    const player = getPlayerFromRegistry(this);
    if (trackId === -1) {
      player?.setTextTrackVisibility(false);
      return;
    }
    player?.setTextTrackVisibility(true);
    player?.setCurrentTextTrack(trackId);
  }
  buildTextTracksSubmenu() {
    if (this.textTracks.length <= 1 || !this.canSetTextTrack) {
      return (h("vm-menu-item", { label: this.i18n.subtitlesOrCc, hint: this.textTracks[this.currentTextTrack]?.label ?? this.i18n.none }));
    }
    return (h("vm-submenu", { label: this.i18n.subtitlesOrCc, hint: this.isTextTrackVisible
        ? this.textTracks[this.currentTextTrack]?.label
        : this.i18n.off },
      h("vm-menu-radio-group", { value: `${!this.isTextTrackVisible ? -1 : this.currentTextTrack}`, onVmCheck: this.onTextTrackSelect.bind(this) }, [h("vm-menu-radio", { label: this.i18n.off, value: "-1" })].concat(this.textTracks.map((track, i) => (h("vm-menu-radio", { label: track.label, value: `${i}` })))))));
  }
  onAudioTrackSelect(event) {
    const radio = event.target;
    const trackId = parseInt(radio.value, 10);
    const player = getPlayerFromRegistry(this);
    player?.setCurrentAudioTrack(trackId);
  }
  buildAudioTracksMenu() {
    if (this.audioTracks.length <= 1 || !this.canSetAudioTrack) {
      return (h("vm-menu-item", { label: this.i18n.audio, hint: this.audioTracks[this.currentAudioTrack]?.label ?? this.i18n.default }));
    }
    return (h("vm-submenu", { label: this.i18n.audio, hint: this.audioTracks[this.currentAudioTrack]?.label },
      h("vm-menu-radio-group", { value: `${this.currentAudioTrack}`, onVmCheck: this.onAudioTrackSelect.bind(this) }, this.audioTracks.map((track, i) => (h("vm-menu-radio", { label: track.label, value: `${i}` }))))));
  }
  render() {
    return (h("vm-settings", { pin: this.pin },
      this.buildAudioTracksMenu(),
      this.buildPlaybackRateSubmenu(),
      this.buildPlaybackQualitySubmenu(),
      this.isVideoView && this.buildTextTracksSubmenu(),
      h("slot", null)));
  }
  static get is() { return "vm-default-settings"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["default-settings.css"]
  }; }
  static get styleUrls() { return {
    "$": ["default-settings.css"]
  }; }
  static get properties() { return {
    "pin": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight'",
        "resolved": "\"bottomLeft\" | \"bottomRight\" | \"topLeft\" | \"topRight\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Pins the settings to the defined position inside the video player. This has no effect when\nthe view is of type `audio`, it will always be `bottomRight`."
      },
      "attribute": "pin",
      "reflect": true,
      "defaultValue": "'bottomRight'"
    },
    "i18n": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['i18n']",
        "resolved": "Translation | { [x: string]: string; }",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "defaultValue": "{}"
    },
    "playbackReady": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['playbackReady']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "playback-ready",
      "reflect": false,
      "defaultValue": "false"
    },
    "playbackRate": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['playbackRate']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "playback-rate",
      "reflect": false,
      "defaultValue": "1"
    },
    "playbackRates": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['playbackRates']",
        "resolved": "number[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "defaultValue": "[1]"
    },
    "isVideoView": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['isAudioView']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "is-video-view",
      "reflect": false,
      "defaultValue": "false"
    },
    "playbackQuality": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['playbackQuality']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "playback-quality",
      "reflect": false
    },
    "playbackQualities": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['playbackQualities']",
        "resolved": "string[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "defaultValue": "[]"
    },
    "textTracks": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['textTracks']",
        "resolved": "TextTrack[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "defaultValue": "[]"
    },
    "currentTextTrack": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "current-text-track",
      "reflect": false,
      "defaultValue": "-1"
    },
    "audioTracks": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "PlayerProps['audioTracks']",
        "resolved": "any[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "../../../core/player/PlayerProps"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "defaultValue": "[]"
    },
    "currentAudioTrack": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "current-audio-track",
      "reflect": false,
      "defaultValue": "-1"
    },
    "isTextTrackVisible": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }],
        "text": ""
      },
      "attribute": "is-text-track-visible",
      "reflect": false,
      "defaultValue": "true"
    }
  }; }
  static get states() { return {
    "canSetPlaybackRate": {},
    "canSetPlaybackQuality": {},
    "canSetTextTrack": {},
    "canSetAudioTrack": {}
  }; }
  static get watchers() { return [{
      "propName": "playbackReady",
      "methodName": "onPlaybackReady"
    }, {
      "propName": "audioTracks",
      "methodName": "onAudioTracksChange"
    }, {
      "propName": "playbackReady",
      "methodName": "onAudioTracksChange"
    }, {
      "propName": "textTracks",
      "methodName": "onTextTracksChange"
    }, {
      "propName": "playbackReady",
      "methodName": "onTextTracksChange"
    }]; }
}
