// devextreme/ui/scheduler/constants.js
/**
 * DevExtreme (ui/scheduler/constants.js)
 * Version: 20.2.6
 * Build date: Tue Mar 16 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = exports.APPOINTMENT_DRAG_SOURCE_CLASS = exports.DIRECTION_APPOINTMENT_CLASSES = exports.REDUCED_APPOINTMENT_PARTS_CLASSES = exports.ALL_DAY_APPOINTMENT_CLASS = exports.EMPTY_APPOINTMENT_CLASS = exports.RECURRENCE_APPOINTMENT_CLASS = exports.REDUCED_APPOINTMENT_ICON = exports.REDUCED_APPOINTMENT_CLASS = exports.APPOINTMENT_SETTINGS_KEY = exports.LIST_ITEM_CLASS = exports.FIXED_CONTAINER_CLASS = exports.LIST_ITEM_DATA_KEY = void 0;
var LIST_ITEM_DATA_KEY = "dxListItemData";
exports.LIST_ITEM_DATA_KEY = LIST_ITEM_DATA_KEY;
var FIXED_CONTAINER_CLASS = "dx-scheduler-fixed-appointments";
exports.FIXED_CONTAINER_CLASS = FIXED_CONTAINER_CLASS;
var LIST_ITEM_CLASS = "dx-list-item";
exports.LIST_ITEM_CLASS = LIST_ITEM_CLASS;
var APPOINTMENT_SETTINGS_KEY = "dxAppointmentSettings";
exports.APPOINTMENT_SETTINGS_KEY = APPOINTMENT_SETTINGS_KEY;
var REDUCED_APPOINTMENT_CLASS = "dx-scheduler-appointment-reduced";
exports.REDUCED_APPOINTMENT_CLASS = REDUCED_APPOINTMENT_CLASS;
var REDUCED_APPOINTMENT_ICON = "dx-scheduler-appointment-reduced-icon";
exports.REDUCED_APPOINTMENT_ICON = REDUCED_APPOINTMENT_ICON;
var RECURRENCE_APPOINTMENT_CLASS = "dx-scheduler-appointment-recurrence";
exports.RECURRENCE_APPOINTMENT_CLASS = RECURRENCE_APPOINTMENT_CLASS;
var EMPTY_APPOINTMENT_CLASS = "dx-scheduler-appointment-empty";
exports.EMPTY_APPOINTMENT_CLASS = EMPTY_APPOINTMENT_CLASS;
var ALL_DAY_APPOINTMENT_CLASS = "dx-scheduler-all-day-appointment";
exports.ALL_DAY_APPOINTMENT_CLASS = ALL_DAY_APPOINTMENT_CLASS;
var REDUCED_APPOINTMENT_PARTS_CLASSES = {
    head: "dx-scheduler-appointment-head",
    body: "dx-scheduler-appointment-body",
    tail: "dx-scheduler-appointment-tail"
};
exports.REDUCED_APPOINTMENT_PARTS_CLASSES = REDUCED_APPOINTMENT_PARTS_CLASSES;
var DIRECTION_APPOINTMENT_CLASSES = {
    horizontal: "dx-scheduler-appointment-horizontal",
    vertical: "dx-scheduler-appointment-vertical"
};
exports.DIRECTION_APPOINTMENT_CLASSES = DIRECTION_APPOINTMENT_CLASSES;
var APPOINTMENT_DRAG_SOURCE_CLASS = "dx-scheduler-appointment-drag-source";
exports.APPOINTMENT_DRAG_SOURCE_CLASS = APPOINTMENT_DRAG_SOURCE_CLASS;
var AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = "dx-scheduler-last-in-date-agenda-appointment";
exports.AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS = AGENDA_LAST_IN_DATE_APPOINTMENT_CLASS;
