// @vime/core/dist/esm/LazyLoader-6a189087.js
import { m as isNullOrUndefined } from './withComponentRegistry-892f231d.js';
import { I as IS_CLIENT } from './support-cbf90a90.js';

class LazyLoader {
  constructor(el, attributes, onLoad) {
    this.el = el;
    this.attributes = attributes;
    this.onLoad = onLoad;
    this.hasLoaded = false;
    if (isNullOrUndefined(this.el))
      return;
    this.intersectionObs = this.canObserveIntersection()
      ? new IntersectionObserver(this.onIntersection.bind(this))
      : undefined;
    this.mutationObs = this.canObserveMutations()
      ? new MutationObserver(this.onMutation.bind(this))
      : undefined;
    this.mutationObs?.observe(this.el, {
      childList: true,
      subtree: true,
      attributeFilter: this.attributes,
    });
    this.lazyLoad();
  }
  didLoad() {
    return this.hasLoaded;
  }
  destroy() {
    this.intersectionObs?.disconnect();
    this.mutationObs?.disconnect();
  }
  canObserveIntersection() {
    return IS_CLIENT && window.IntersectionObserver;
  }
  canObserveMutations() {
    return IS_CLIENT && window.MutationObserver;
  }
  lazyLoad() {
    if (this.canObserveIntersection()) {
      this.intersectionObs?.observe(this.el);
    }
    else {
      this.load();
    }
  }
  onIntersection(entries) {
    entries.forEach((entry) => {
      if (entry.intersectionRatio > 0 || entry.isIntersecting) {
        this.load();
        this.intersectionObs.unobserve(entry.target);
      }
    });
  }
  onMutation() {
    if (this.hasLoaded)
      this.load();
  }
  getLazyElements() {
    const root = !isNullOrUndefined(this.el.shadowRoot)
      ? this.el.shadowRoot
      : this.el;
    return root.querySelectorAll('.lazy');
  }
  load() {
    window.requestAnimationFrame(() => {
      this.getLazyElements().forEach(this.loadEl.bind(this));
    });
  }
  loadEl(el) {
    this.intersectionObs?.unobserve(el);
    this.hasLoaded = true;
    this.onLoad?.(el);
  }
}

export { LazyLoader as L };
