// @vime/core/dist/esm/vm-default-ui.entry.js
import { r as registerInstance, h } from './index-d8f97135.js';
import { w as withComponentRegistry } from './withComponentRegistry-892f231d.js';

const defaultUiCss = ":host{display:contents;pointer-events:none}";

const DefaultUI = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Whether clicking the player should not toggle playback.
     */
    this.noClickToPlay = false;
    /**
     * Whether double clicking the player should not toggle fullscreen mode.
     */
    this.noDblClickFullscreen = false;
    /**
     * Whether the custom captions UI should not be loaded.
     */
    this.noCaptions = false;
    /**
     * Whether the custom poster UI should not be loaded.
     */
    this.noPoster = false;
    /**
     * Whether the custom spinner UI should not be loaded.
     */
    this.noSpinner = false;
    /**
     * Whether the custom default controls should not be loaded.
     */
    this.noControls = false;
    /**
     * Whether the custom default settings menu should not be loaded.
     */
    this.noSettings = false;
    /**
     * Whether the default loading screen should not be loaded.
     */
    this.noLoadingScreen = false;
    withComponentRegistry(this);
  }
  render() {
    return (h("vm-ui", null, !this.noClickToPlay && h("vm-click-to-play", null), !this.noDblClickFullscreen && h("vm-dbl-click-fullscreen", null), !this.noCaptions && h("vm-captions", null), !this.noPoster && h("vm-poster", null), !this.noSpinner && h("vm-spinner", null), !this.noLoadingScreen && h("vm-loading-screen", null), !this.noControls && h("vm-default-controls", null), !this.noSettings && h("vm-default-settings", null), h("slot", null)));
  }
};
DefaultUI.style = defaultUiCss;

export { DefaultUI as vm_default_ui };
