// @vime/core/dist/collection/components/core/player/player.js
import { Component, Element, Event, h, Host, Listen, Method, Prop, State, Watch, writeTask, } from '@stencil/core';
import { Universe } from 'stencil-wormhole';
import { MediaType } from './MediaType';
import { isUndefined, isString } from '../../../utils/unit';
import { initialState } from './PlayerProps';
import { ViewType } from './ViewType';
import { canAutoplay, onTouchInputChange, IS_IOS, canRotateScreen, onMobileChange, } from '../../../utils/support';
import { Fullscreen } from './fullscreen/Fullscreen';
import { en } from './lang/en';
import { Disposal } from '../../../utils/Disposal';
import { Logger } from './PlayerLogger';
import { autopause, withAutopause } from './withAutopause';
import { withProviderHost } from '../../providers/ProviderConnect';
import { withPlayerEvents } from './withPlayerEvents';
import { withPlayerScheduler } from './withPlayerScheduler';
import { isComponentRegistered, withComponentRegistrar, } from './withComponentRegistry';
import { withFindPlayer } from './findPlayer';
let idCount = 0;
/**
 * @slot - Used to pass in providers, plugins and UI components.
 */
export class Player {
  constructor() {
    this.disposal = new Disposal();
    /**
     * ------------------------------------------------------
     * Props
     * ------------------------------------------------------
     */
    /** @internal @readonly */
    this.logger = new Logger();
    /** @inheritDoc */
    this.icons = 'vime';
    /** @inheritDoc */
    this.paused = true;
    /** @inheritDoc @readonly */
    this.playing = false;
    /** @inheritDoc @readonly */
    this.duration = -1;
    /** @inheritDoc */
    this.currentTime = 0;
    /** @inheritDoc */
    this.autoplay = false;
    /** @inheritDoc @readonly */
    this.ready = false;
    /** @inheritDoc @readonly */
    this.playbackReady = false;
    /** @inheritDoc */
    this.loop = false;
    /** @inheritDoc */
    this.muted = false;
    /** @inheritDoc @readonly */
    this.buffered = 0;
    /** @inheritDoc */
    this.playbackRate = 1;
    this.lastRateCheck = 1;
    /** @inheritDoc @readonly */
    this.playbackRates = [1];
    /** @inheritDoc @readonly */
    this.playbackQualities = [];
    /** @inheritDoc @readonly */
    this.seeking = false;
    /** @inheritDoc */
    this.debug = false;
    /** @inheritDoc @readonly */
    this.playbackStarted = false;
    /** @inheritDoc @readonly */
    this.playbackEnded = false;
    /** @inheritDoc @readonly */
    this.buffering = false;
    /** @inheritDoc */
    this.controls = false;
    /** @inheritDoc */
    this.isControlsActive = false;
    /** @inheritDoc @readonly */
    this.isSettingsActive = false;
    /** @inheritDoc */
    this.volume = 50;
    /** @inheritDoc @readonly */
    this.isFullscreenActive = false;
    /** @inheritDoc */
    this.aspectRatio = '16:9';
    /** @inheritDoc @readonly */
    this.isAudioView = false;
    /** @inheritDoc @readonly */
    this.isVideoView = false;
    /** @inheritDoc @readonly */
    this.isAudio = false;
    /** @inheritDoc @readonly */
    this.isVideo = false;
    /** @inheritDoc @readonly */
    this.isLive = false;
    /** @inheritDoc @readonly */
    this.isMobile = false;
    /** @inheritDoc @readonly */
    this.isTouch = false;
    /** @inheritDoc @readonly */
    this.isPiPActive = false;
    /** @inheritDoc @readonly */
    this.textTracks = [];
    /** @inheritDoc @readonly */
    this.currentTextTrack = -1;
    /** @inheritDoc @readonly */
    this.isTextTrackVisible = true;
    /** @inheritDoc */
    this.shouldRenderNativeTextTracks = true;
    /** @inheritDoc @readonly */
    this.audioTracks = [];
    /** @inheritDoc @readonly */
    this.currentAudioTrack = -1;
    /** @inheritDoc */
    this.autopause = true;
    /** @inheritDoc */
    this.playsinline = false;
    /** @inheritDoc */
    this.language = 'en';
    /** @inheritDoc */
    this.translations = { en };
    /** @inheritDoc @readonly */
    this.languages = ['en'];
    /** @inheritDoc @readonly */
    this.i18n = en;
    withFindPlayer(this);
    withComponentRegistrar(this);
    withAutopause(this);
    withProviderHost(this);
    withPlayerEvents(this);
    this.safeAdapterCall = withPlayerScheduler(this);
  }
  get adapter() {
    return this.provider?.getAdapter();
  }
  onContainerChange() {
    this.fullscreen?.destroy();
    if (isUndefined(this.container))
      return;
    this.fullscreen = new Fullscreen(this.container, (isActive) => {
      this.isFullscreenActive = isActive;
      this.rotateDevice();
    });
  }
  onPausedChange() {
    if (this.paused) {
      this.playing = false;
    }
    else {
      autopause(this);
    }
    this.safeAdapterCall('paused', !this.paused ? 'play' : 'pause');
  }
  onDurationChange() {
    this.isLive = this.duration === Infinity;
  }
  onCurrentTimeChange() {
    const duration = this.playbackReady ? this.duration : Infinity;
    this.currentTime = Math.max(0, Math.min(this.currentTime, duration));
    this.safeAdapterCall('currentTime', 'setCurrentTime');
  }
  onPlaybackReadyChange() {
    if (!this.ready)
      this.ready = true;
  }
  onMutedChange() {
    this.safeAdapterCall('muted', 'setMuted');
  }
  async onPlaybackRateChange(newRate, prevRate) {
    if (newRate === this.lastRateCheck)
      return;
    if (!(await (await this.adapter)?.canSetPlaybackRate?.())) {
      this.logger.log('provider cannot change `playbackRate`.');
      this.lastRateCheck = prevRate;
      this.playbackRate = prevRate;
      return;
    }
    if (!this.playbackRates.includes(newRate)) {
      this.logger.log(`invalid \`playbackRate\` of ${newRate}, ` +
        `valid values are [${this.playbackRates.join(', ')}]`);
      this.lastRateCheck = prevRate;
      this.playbackRate = prevRate;
      return;
    }
    this.lastRateCheck = newRate;
    this.safeAdapterCall('playbackRate', 'setPlaybackRate');
  }
  async onPlaybackQualityChange(newQuality, prevQuality) {
    if (isUndefined(newQuality) || newQuality === this.lastQualityCheck)
      return;
    if (!(await (await this.adapter)?.canSetPlaybackQuality?.())) {
      this.logger.log('provider cannot change `playbackQuality`.');
      this.lastQualityCheck = prevQuality;
      this.playbackQuality = prevQuality;
      return;
    }
    if (!this.playbackQualities.includes(newQuality)) {
      this.logger.log(`invalid \`playbackQuality\` of ${newQuality}, ` +
        `valid values are [${this.playbackQualities.join(', ')}]`);
      this.lastQualityCheck = prevQuality;
      this.playbackQuality = prevQuality;
      return;
    }
    this.lastQualityCheck = newQuality;
    this.safeAdapterCall('playbackQuality', 'setPlaybackQuality');
  }
  onDebugChange() {
    this.logger.silent = !this.debug;
  }
  async onVolumeChange() {
    this.volume = Math.max(0, Math.min(this.volume, 100));
    this.safeAdapterCall('volume', 'setVolume');
  }
  onViewTypeChange() {
    this.isAudioView = this.viewType === ViewType.Audio;
    this.isVideoView = this.viewType === ViewType.Video;
  }
  onMediaTypeChange() {
    this.isAudio = this.mediaType === MediaType.Audio;
    this.isVideo = this.mediaType === MediaType.Video;
  }
  onLanguageChange(_, prevLanguage) {
    if (!this.languages.includes(this.language)) {
      this.logger.log(`invalid \`language\` of ${this.language}, ` +
        `valid values are [${this.languages.join(', ')}]`);
      this.language = prevLanguage;
      return;
    }
    this.i18n = this.translations[this.language];
  }
  onTranslationsChange() {
    Object.assign(this.translations, { en });
    this.languages = Object.keys(this.translations);
    this.i18n = this.translations[this.language];
  }
  onError(event) {
    this.logger.warn(event.detail);
  }
  /**
   * ------------------------------------------------------
   * Methods
   * ------------------------------------------------------
   */
  /** @inheritDoc */
  async getProvider() {
    return this.provider;
  }
  /** @internal */
  async getAdapter() {
    return this.adapter;
  }
  /** @inheritDoc */
  async play() {
    return (await this.adapter)?.play();
  }
  /** @inheritDoc */
  async pause() {
    return (await this.adapter)?.pause();
  }
  /** @inheritDoc */
  async canPlay(type) {
    return (await this.adapter)?.canPlay(type) ?? false;
  }
  /** @inheritDoc */
  async canAutoplay() {
    return canAutoplay();
  }
  /** @inheritDoc */
  async canMutedAutoplay() {
    return canAutoplay(true);
  }
  /** @inheritDoc */
  async canSetPlaybackRate() {
    return (await this.adapter)?.canSetPlaybackRate?.() ?? false;
  }
  /** @inheritDoc */
  async canSetPlaybackQuality() {
    return (await this.adapter)?.canSetPlaybackQuality?.() ?? false;
  }
  /** @inheritDoc */
  async canSetFullscreen() {
    return (this.fullscreen.isSupported ||
      ((await this.adapter)?.canSetFullscreen?.() ?? false));
  }
  /** @inheritDoc */
  async enterFullscreen(options) {
    if (!this.isVideoView)
      throw Error('Cannot enter fullscreen on an audio player view.');
    if (this.fullscreen.isSupported)
      return this.fullscreen.enterFullscreen(options);
    if (await (await this.adapter)?.canSetFullscreen?.()) {
      return (await this.adapter)?.enterFullscreen?.(options);
    }
    throw Error('Fullscreen API is not available.');
  }
  /** @inheritDoc */
  async exitFullscreen() {
    if (this.fullscreen.isSupported)
      return this.fullscreen.exitFullscreen();
    return (await this.adapter)?.exitFullscreen?.();
  }
  /** @inheritDoc */
  async canSetPiP() {
    return (await this.adapter)?.canSetPiP?.() ?? false;
  }
  /** @inheritDoc */
  async enterPiP() {
    if (!this.isVideoView)
      throw Error('Cannot enter PiP mode on an audio player view.');
    if (!(await this.canSetPiP()))
      throw Error('Picture-in-Picture API is not available.');
    return (await this.adapter)?.enterPiP?.();
  }
  /** @inheritDoc */
  async exitPiP() {
    return (await this.adapter)?.exitPiP?.();
  }
  /** @inheritDoc */
  async canSetAudioTrack() {
    return !isUndefined((await this.adapter)?.setCurrentAudioTrack);
  }
  /** @inheritDoc */
  async setCurrentAudioTrack(trackId) {
    (await this.adapter)?.setCurrentAudioTrack?.(trackId);
  }
  /** @inheritDoc */
  async canSetTextTrack() {
    return !isUndefined((await this.adapter)?.setCurrentTextTrack);
  }
  /** @inheritDoc */
  async setCurrentTextTrack(trackId) {
    (await this.adapter)?.setCurrentTextTrack?.(trackId);
  }
  /** @inheritDoc */
  async canSetTextTrackVisibility() {
    return !isUndefined((await this.adapter)?.setTextTrackVisibility);
  }
  /** @inheritDoc */
  async setTextTrackVisibility(isVisible) {
    (await this.adapter)?.setTextTrackVisibility?.(isVisible);
  }
  /** @inheritDoc */
  async extendLanguage(language, translation) {
    const translations = {
      ...this.translations,
      [language]: {
        ...(this.translations[language] ?? {}),
        ...translation,
      },
    };
    this.translations = translations;
  }
  connectedCallback() {
    this.onPausedChange();
    this.onCurrentTimeChange();
    this.onVolumeChange();
    this.onMutedChange();
    this.onDebugChange();
    this.onContainerChange();
    this.onTranslationsChange();
    this.onLanguageChange(this.language, initialState.language);
    this.disposal.add(onMobileChange((isMobile) => {
      this.isMobile = isMobile;
    }));
    this.disposal.add(onTouchInputChange((isTouch) => {
      this.isTouch = isTouch;
    }));
  }
  componentWillLoad() {
    Universe.create(this, this.getPlayerState());
  }
  disconnectedCallback() {
    this.fullscreen?.destroy();
    this.disposal.empty();
  }
  async rotateDevice() {
    if (!this.isMobile || !canRotateScreen())
      return;
    try {
      if (this.isFullscreenActive) {
        await window.screen.orientation.lock('landscape');
      }
      else {
        await window.screen.orientation.unlock();
      }
    }
    catch (err) {
      this.vmError.emit(err);
    }
  }
  getPlayerState() {
    const state = {};
    const props = Object.keys(initialState);
    for (let i = 0; i < props.length; i += 1) {
      state[props[i]] = this[props[i]];
    }
    return state;
  }
  calcAspectRatio() {
    const [width, height] = /\d{1,2}:\d{1,2}/.test(this.aspectRatio)
      ? this.aspectRatio.split(':')
      : [16, 9];
    return (100 / Number(width)) * Number(height);
  }
  /**
   * Returns the inner container.
   */
  async getContainer() {
    return this.container;
  }
  /** @internal Exposed for E2E testing. */
  async callAdapter(method, value) {
    return (await this.adapter)[method](value);
  }
  hasCustomControls() {
    return isComponentRegistered(this, 'vm-controls');
  }
  genId() {
    const id = this.host?.id;
    if (isString(id) && id.length > 0)
      return id;
    idCount += 1;
    return `vm-player-${idCount}`;
  }
  render() {
    const label = `${this.isAudioView ? 'Audio Player' : 'Video Player'}` +
      `${!isUndefined(this.mediaTitle) ? ` - ${this.mediaTitle}` : ''}`;
    const canShowCustomUI = !IS_IOS ||
      !this.isVideoView ||
      (this.playsinline && !this.isFullscreenActive);
    if (!canShowCustomUI) {
      this.controls = true;
    }
    const isIdle = canShowCustomUI &&
      this.hasCustomControls() &&
      this.isVideoView &&
      !this.paused &&
      !this.isControlsActive;
    const isBlockerVisible = !this.controls && canShowCustomUI && this.isVideoView;
    return (h(Host, { id: this.genId(), idle: isIdle, mobile: this.isMobile, touch: this.isTouch, live: this.isLive, audio: this.isAudioView, video: this.isVideoView, pip: this.isPiPActive, fullscreen: this.isFullscreenActive },
      h("div", { "aria-label": label, "aria-hidden": !this.ready ? 'true' : 'false', "aria-busy": !this.playbackReady ? 'true' : 'false', class: {
          player: true,
          idle: isIdle,
          audio: this.isAudioView,
          video: this.isVideoView,
          fullscreen: this.isFullscreenActive,
        }, style: {
          paddingBottom: this.isVideoView
            ? `${this.calcAspectRatio()}%`
            : undefined,
        }, ref: (el) => {
          writeTask(() => {
            this.container = el;
          });
        } },
        isBlockerVisible && h("div", { class: "blocker" }),
        h(Universe.Provider, { state: this.getPlayerState() },
          h("slot", null)))));
  }
  static get is() { return "vm-player"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["player.css"]
  }; }
  static get styleUrls() { return {
    "$": ["player.css"]
  }; }
  static get properties() { return {
    "logger": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "Logger",
        "resolved": "Logger",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "internal"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": ""
      },
      "defaultValue": "new Logger()"
    },
    "theme": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "This property has no role other than scoping CSS selectors."
      },
      "attribute": "theme",
      "reflect": true
    },
    "icons": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "The default icon library to be used throughout the player. You can use a predefined\nicon library such as vime, material, remix or boxicons. If you'd like to provide your own\nsee the `<vm-icon-library>` component. Remember to pass in the name of your icon library here."
      },
      "attribute": "icons",
      "reflect": true,
      "defaultValue": "'vime'"
    },
    "paused": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether playback should be paused. Defaults to `true` if no media has loaded or playback has\nnot started. Setting this to `true` will begin/resume playback."
      },
      "attribute": "paused",
      "reflect": false,
      "defaultValue": "true"
    },
    "playing": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether media is actively playing back. Defaults to `false` if no media has\nloaded or playback has not started."
      },
      "attribute": "playing",
      "reflect": false,
      "defaultValue": "false"
    },
    "duration": {
      "type": "number",
      "mutable": true,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "A `double` indicating the total playback length of the media in seconds. Defaults\nto `-1` if no media has been loaded. If the media is being streamed live then the duration is\nequal to `Infinity`."
      },
      "attribute": "duration",
      "reflect": false,
      "defaultValue": "-1"
    },
    "mediaTitle": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The title of the current media. Defaults to `undefined` if no media has been\nloaded."
      },
      "attribute": "media-title",
      "reflect": false
    },
    "currentProvider": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "Provider",
        "resolved": "Provider.Audio | Provider.Dailymotion | Provider.Dash | Provider.HLS | Provider.Video | Provider.Vimeo | Provider.YouTube | undefined",
        "references": {
          "Provider": {
            "location": "import",
            "path": "../../providers/Provider"
          }
        }
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The current provider name whose responsible for loading and playing media.\nDefaults to `undefined` when no provider has been loaded."
      },
      "attribute": "current-provider",
      "reflect": false
    },
    "currentSrc": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The absolute URL of the media resource that has been chosen. Defaults to\n`undefined` if no media has been loaded."
      },
      "attribute": "current-src",
      "reflect": false
    },
    "currentPoster": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The absolute URL of the poster for the current media resource. Defaults to\n`undefined` if no media/poster has been loaded."
      },
      "attribute": "current-poster",
      "reflect": false
    },
    "currentTime": {
      "type": "number",
      "mutable": true,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "A `double` indicating the current playback time in seconds. Defaults to `0` if the media has\nnot started to play and has not seeked. Setting this value seeks the media to the new\ntime. The value can be set to a minimum of `0` and maximum of the total length of the\nmedia (indicated by the duration prop)."
      },
      "attribute": "current-time",
      "reflect": false,
      "defaultValue": "0"
    },
    "autoplay": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether playback should automatically begin playing once the media is ready to do so. This\nwill only work if the browsers `autoplay` policies have been satisfied. It'll generally work\nif the player is muted, or the user frequently interacts with your site. You can check\nif it's possible to autoplay via the `canAutoplay()` or `canMutedAutoplay()` methods.\nDepending on the provider, changing this prop may cause the player to completely reset."
      },
      "attribute": "autoplay",
      "reflect": false,
      "defaultValue": "false"
    },
    "ready": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player has loaded and is ready to be interacted with."
      },
      "attribute": "ready",
      "reflect": true,
      "defaultValue": "false"
    },
    "playbackReady": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether media is ready for playback to begin."
      },
      "attribute": "playback-ready",
      "reflect": false,
      "defaultValue": "false"
    },
    "loop": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether media should automatically start playing from the beginning every time it ends."
      },
      "attribute": "loop",
      "reflect": false,
      "defaultValue": "false"
    },
    "muted": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether the audio is muted or not."
      },
      "attribute": "muted",
      "reflect": false,
      "defaultValue": "false"
    },
    "buffered": {
      "type": "number",
      "mutable": true,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The length of the media in seconds that has been downloaded by the browser."
      },
      "attribute": "buffered",
      "reflect": false,
      "defaultValue": "0"
    },
    "playbackRate": {
      "type": "number",
      "mutable": true,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "A `double` indicating the rate at which media is being played back. If the value is `<1` then\nplayback is slowed down; if `>1` then playback is sped up. Defaults to `1`. The playback rate\ncan only be set to a rate found in the `playbackRates` prop. Some providers may not\nallow changing the playback rate, you can check if it's possible via `canSetPlaybackRate()`."
      },
      "attribute": "playback-rate",
      "reflect": false,
      "defaultValue": "1"
    },
    "playbackRates": {
      "type": "unknown",
      "mutable": true,
      "complexType": {
        "original": "number[]",
        "resolved": "number[]",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The playback rates available for the current media."
      },
      "defaultValue": "[1]"
    },
    "playbackQuality": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string | undefined",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Indicates the quality of the media. The value will differ between audio and video. For audio\nthis might be some combination of the encoding format (AAC, MP3), bitrate in kilobits per\nsecond (kbps) and sample rate in kilohertz (kHZ). For video this will be the number of vertical\npixels it supports. For example, if the video has a resolution of `1920x1080` then the quality\nwill return `1080p`. Defaults to `undefined` which you can interpret as the quality is unknown.\nThe value can also be `Auto` for adaptive bit streams (ABR), where the provider can\nautomatically manage the playback quality. The quality can only be set to a quality found\nin the `playbackQualities` prop. Some providers may not allow changing the quality, you can\ncheck if it's possible via `canSetPlaybackQuality()`."
      },
      "attribute": "playback-quality",
      "reflect": false
    },
    "playbackQualities": {
      "type": "unknown",
      "mutable": true,
      "complexType": {
        "original": "string[]",
        "resolved": "string[]",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The media qualities available for the current media."
      },
      "defaultValue": "[]"
    },
    "seeking": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player is in the process of seeking to a new time position."
      },
      "attribute": "seeking",
      "reflect": false,
      "defaultValue": "false"
    },
    "debug": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether the player is in debug mode and should `console.x` information about\nits internal state."
      },
      "attribute": "debug",
      "reflect": false,
      "defaultValue": "false"
    },
    "playbackStarted": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the media has initiated playback. In other words it will be true if\n`currentTime > 0`."
      },
      "attribute": "playback-started",
      "reflect": false,
      "defaultValue": "false"
    },
    "playbackEnded": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether media playback has reached the end. In other words it'll be true if\n`currentTime === duration`."
      },
      "attribute": "playback-ended",
      "reflect": false,
      "defaultValue": "false"
    },
    "buffering": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether playback has temporarily stopped because of a lack of temporary data."
      },
      "attribute": "buffering",
      "reflect": false,
      "defaultValue": "false"
    },
    "controls": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Indicates whether a user interface should be shown for controlling the resource. Set this to\n`false` when you want to provide your own custom controls, and `true` if you want the current\nprovider to supply its own default controls. Depending on the provider, changing this prop\nmay cause the player to completely reset."
      },
      "attribute": "controls",
      "reflect": false,
      "defaultValue": "false"
    },
    "isControlsActive": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether the controls are currently visible. This is currently only supported by custom\ncontrols."
      },
      "attribute": "is-controls-active",
      "reflect": false,
      "defaultValue": "false"
    },
    "isSettingsActive": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the settings menu has been opened and is currently visible. This is\ncurrently only supported by custom settings."
      },
      "attribute": "is-settings-active",
      "reflect": false,
      "defaultValue": "false"
    },
    "volume": {
      "type": "number",
      "mutable": true,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "An `int` between `0` (silent) and `100` (loudest) indicating the audio volume."
      },
      "attribute": "volume",
      "reflect": false,
      "defaultValue": "50"
    },
    "isFullscreenActive": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player is currently in fullscreen mode."
      },
      "attribute": "is-fullscreen-active",
      "reflect": false,
      "defaultValue": "false"
    },
    "aspectRatio": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "The aspect ratio of the player expressed as `width:height` (`16:9`). This is only applied if\nthe `viewType` is `video` and the player is not in fullscreen mode."
      },
      "attribute": "aspect-ratio",
      "reflect": false,
      "defaultValue": "'16:9'"
    },
    "viewType": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "ViewType",
        "resolved": "ViewType.Audio | ViewType.Video | undefined",
        "references": {
          "ViewType": {
            "location": "import",
            "path": "./ViewType"
          }
        }
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The type of player view that is being used, whether it's an audio player view or\nvideo player view. Normally if the media type is of audio then the view is of type audio, but\nin some cases it might be desirable to show a different view type. For example, when playing\naudio with a poster. This is subject to the provider allowing it. Defaults to `undefined`\nwhen no media has been loaded."
      },
      "attribute": "view-type",
      "reflect": false
    },
    "isAudioView": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current view is of type `audio`, shorthand for\n`viewType === ViewType.Audio`."
      },
      "attribute": "is-audio-view",
      "reflect": false,
      "defaultValue": "false"
    },
    "isVideoView": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current view is of type `video`, shorthand for\n`viewType === ViewType.Video`."
      },
      "attribute": "is-video-view",
      "reflect": false,
      "defaultValue": "false"
    },
    "mediaType": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "MediaType",
        "resolved": "MediaType.Audio | MediaType.Video | undefined",
        "references": {
          "MediaType": {
            "location": "import",
            "path": "./MediaType"
          }
        }
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The type of media that is currently active, whether it's audio or video. Defaults\nto `undefined` when no media has been loaded or the type cannot be determined."
      },
      "attribute": "media-type",
      "reflect": false
    },
    "isAudio": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current media is of type `audio`, shorthand for\n`mediaType === MediaType.Audio`."
      },
      "attribute": "is-audio",
      "reflect": false,
      "defaultValue": "false"
    },
    "isVideo": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current media is of type `video`, shorthand for\n`mediaType === MediaType.Video`."
      },
      "attribute": "is-video",
      "reflect": false,
      "defaultValue": "false"
    },
    "isLive": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current media is being broadcast live (`duration === Infinity`)."
      },
      "attribute": "is-live",
      "reflect": false,
      "defaultValue": "false"
    },
    "isMobile": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player is in mobile mode. This is determined by parsing\n`window.navigator.userAgent`."
      },
      "attribute": "is-mobile",
      "reflect": false,
      "defaultValue": "false"
    },
    "isTouch": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player is in touch mode. This is determined by listening for\nmouse/touch events and toggling this value."
      },
      "attribute": "is-touch",
      "reflect": false,
      "defaultValue": "false"
    },
    "isPiPActive": {
      "type": "boolean",
      "mutable": true,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the player is currently in picture-in-picture mode."
      },
      "attribute": "is-pi-p-active",
      "reflect": false,
      "defaultValue": "false"
    },
    "textTracks": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "never[]",
        "resolved": "never[]",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The text tracks associated with the current media."
      },
      "defaultValue": "[]"
    },
    "currentTextTrack": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Gets the index of the currently active text track. Defaults to `-1` to when\nall text tracks are disabled. If you'd like to set it than see the `setCurrentTextTrack`\nmethod."
      },
      "attribute": "current-text-track",
      "reflect": false,
      "defaultValue": "-1"
    },
    "isTextTrackVisible": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Whether the current text tracks is visible. If you'd like to set it than see\nthe `setTrackTrackVisibility` method."
      },
      "attribute": "is-text-track-visible",
      "reflect": false,
      "defaultValue": "true"
    },
    "shouldRenderNativeTextTracks": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether text tracks should be rendered by native player, set to `false` if using\ncustom display."
      },
      "attribute": "should-render-native-text-tracks",
      "reflect": false,
      "defaultValue": "true"
    },
    "audioTracks": {
      "type": "unknown",
      "mutable": false,
      "complexType": {
        "original": "never[]",
        "resolved": "never[]",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The audio tracks associated with the current media."
      },
      "defaultValue": "[]"
    },
    "currentAudioTrack": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "Gets the index of the currently active audio track. Defaults to `-1` to when\nthe default audio track is used. If you'd like to set it than see the `setCurrentAudioTrack`\nmethod."
      },
      "attribute": "current-audio-track",
      "reflect": false,
      "defaultValue": "-1"
    },
    "autopause": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether the player should automatically pause when another Vime player starts/resumes playback."
      },
      "attribute": "autopause",
      "reflect": false,
      "defaultValue": "true"
    },
    "playsinline": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Whether the video is to be played \"inline\", that is within the element's playback area. Note\nthat setting this to false does not imply that the video will always be played in fullscreen.\nDepending on the provider, changing this prop may cause the player to completely reset."
      },
      "attribute": "playsinline",
      "reflect": false,
      "defaultValue": "false"
    },
    "language": {
      "type": "string",
      "mutable": true,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "The current language of the player. This can be any code defined via the `extendLanguage`\nmethod or the default `en`. It's recommended to use an ISO 639-1 code as that'll be used by\nVime when adding new language defaults in the future."
      },
      "attribute": "language",
      "reflect": false,
      "defaultValue": "'en'"
    },
    "translations": {
      "type": "unknown",
      "mutable": true,
      "complexType": {
        "original": "Record<string, Translation>",
        "resolved": "{ [x: string]: Translation; }",
        "references": {
          "Record": {
            "location": "global"
          },
          "Translation": {
            "location": "import",
            "path": "./lang/Translation"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Contains each language and its respective translation map."
      },
      "defaultValue": "{ en }"
    },
    "languages": {
      "type": "unknown",
      "mutable": true,
      "complexType": {
        "original": "string[]",
        "resolved": "string[]",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "The languages that are currently available. You can add new languages via the\n`extendLanguage` method."
      },
      "defaultValue": "['en']"
    },
    "i18n": {
      "type": "unknown",
      "mutable": true,
      "complexType": {
        "original": "Translation",
        "resolved": "Translation",
        "references": {
          "Translation": {
            "location": "import",
            "path": "./lang/Translation"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }, {
            "text": undefined,
            "name": "readonly"
          }],
        "text": "A dictionary of translations for the current language."
      },
      "defaultValue": "en"
    }
  }; }
  static get states() { return {
    "container": {}
  }; }
  static get events() { return [{
      "method": "vmThemeChange",
      "name": "vmThemeChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `theme` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['theme']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPausedChange",
      "name": "vmPausedChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `paused` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['paused']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPlay",
      "name": "vmPlay",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the media is transitioning from `paused` to `playing`. Event flow: `paused` ->\n`play` -> `playing`. The media starts `playing` once enough content has buffered to\nbegin/resume playback."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmPlayingChange",
      "name": "vmPlayingChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `playing` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['playing']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmSeekingChange",
      "name": "vmSeekingChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `seeking` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['seeking']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmSeeked",
      "name": "vmSeeked",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted directly after the player has successfully transitioned/seeked to a new time position.\nEvent flow: `seeking` -> `seeked`."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmBufferingChange",
      "name": "vmBufferingChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `buffering` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['buffering']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmDurationChange",
      "name": "vmDurationChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `duration` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['duration']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmCurrentTimeChange",
      "name": "vmCurrentTimeChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentTime` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentTime']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmReady",
      "name": "vmReady",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the player has loaded and is ready to be interacted with."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmPlaybackReady",
      "name": "vmPlaybackReady",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the media is ready to begin playback. The following props are guaranteed to be\ndefined when this fires: `mediaTitle`, `currentSrc`, `currentPoster`, `duration`, `mediaType`,\n`viewType`."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmPlaybackStarted",
      "name": "vmPlaybackStarted",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the media initiates playback."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmPlaybackEnded",
      "name": "vmPlaybackEnded",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when playback reaches the end of the media."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmBufferedChange",
      "name": "vmBufferedChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `buffered` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['buffered']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmError",
      "name": "vmError",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when an any error has occurred within the player."
      },
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      }
    }, {
      "method": "vmLoadStart",
      "name": "vmLoadStart",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the provider starts loading a media resource."
      },
      "complexType": {
        "original": "void",
        "resolved": "void",
        "references": {}
      }
    }, {
      "method": "vmCurrentProviderChange",
      "name": "vmCurrentProviderChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentProvider` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentProvider']",
        "resolved": "Provider.Audio | Provider.Dailymotion | Provider.Dash | Provider.HLS | Provider.Video | Provider.Vimeo | Provider.YouTube | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmCurrentSrcChange",
      "name": "vmCurrentSrcChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentSrc` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentSrc']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmCurrentPosterChange",
      "name": "vmCurrentPosterChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentPoster` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentPoster']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmMediaTitleChange",
      "name": "vmMediaTitleChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `mediaTitle` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['mediaTitle']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmControlsChange",
      "name": "vmControlsChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isControlsActive` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isControlsActive']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPlaybackRateChange",
      "name": "vmPlaybackRateChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `playbackRate` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['playbackRate']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPlaybackRatesChange",
      "name": "vmPlaybackRatesChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `playbackRates` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['playbackRates']",
        "resolved": "number[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPlaybackQualityChange",
      "name": "vmPlaybackQualityChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `playbackQuality` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['playbackQuality']",
        "resolved": "string | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPlaybackQualitiesChange",
      "name": "vmPlaybackQualitiesChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `playbackQualities` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['playbackQualities']",
        "resolved": "string[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmMutedChange",
      "name": "vmMutedChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `muted` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['muted']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmVolumeChange",
      "name": "vmVolumeChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `volume` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['volume']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmViewTypeChange",
      "name": "vmViewTypeChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `viewType` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['viewType']",
        "resolved": "ViewType.Audio | ViewType.Video | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmMediaTypeChange",
      "name": "vmMediaTypeChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `mediaType` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['mediaType']",
        "resolved": "MediaType.Audio | MediaType.Video | undefined",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmLiveChange",
      "name": "vmLiveChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isLive` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isLive']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmTouchChange",
      "name": "vmTouchChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isTouch` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isTouch']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmLanguageChange",
      "name": "vmLanguageChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `language` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['language']",
        "resolved": "string",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmI18nChange",
      "name": "vmI18nChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `i18n` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['i18n']",
        "resolved": "Translation | { [x: string]: string; }",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmTranslationsChange",
      "name": "vmTranslationsChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `translations` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['translations']",
        "resolved": "{ [x: string]: Translation; }",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmLanguagesChange",
      "name": "vmLanguagesChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `languages` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['languages']",
        "resolved": "string[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmFullscreenChange",
      "name": "vmFullscreenChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isFullscreenActive` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isFullscreenActive']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmPiPChange",
      "name": "vmPiPChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isPiPActive` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isPiPActive']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmTextTracksChange",
      "name": "vmTextTracksChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `textTracks` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['textTracks']",
        "resolved": "TextTrack[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmCurrentTextTrackChange",
      "name": "vmCurrentTextTrackChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentTextTrack` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentTextTrack']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmTextTrackVisibleChange",
      "name": "vmTextTrackVisibleChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `isTextTrackVisible` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['isTextTrackVisible']",
        "resolved": "boolean",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmAudioTracksChange",
      "name": "vmAudioTracksChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `audioTracks` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['audioTracks']",
        "resolved": "any[]",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }, {
      "method": "vmCurrentAudioTrackChange",
      "name": "vmCurrentAudioTrackChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [{
            "text": undefined,
            "name": "inheritDoc"
          }],
        "text": "Emitted when the `currentAudioTrack` prop changes value."
      },
      "complexType": {
        "original": "PlayerProps['currentAudioTrack']",
        "resolved": "number",
        "references": {
          "PlayerProps": {
            "location": "import",
            "path": "./PlayerProps"
          }
        }
      }
    }]; }
  static get methods() { return {
    "getProvider": {
      "complexType": {
        "signature": "<InternalPlayerType = any>() => Promise<AdapterHost<InternalPlayerType> | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "AdapterHost": {
            "location": "import",
            "path": "../../providers/MediaProvider"
          },
          "InternalPlayerType": {
            "location": "global"
          }
        },
        "return": "Promise<AdapterHost<InternalPlayerType> | undefined>"
      },
      "docs": {
        "text": "Returns the current media provider.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "getAdapter": {
      "complexType": {
        "signature": "<InternalPlayerType = any>() => Promise<MediaProviderAdapter<InternalPlayerType> | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "MediaProviderAdapter": {
            "location": "import",
            "path": "../../providers/MediaProvider"
          },
          "InternalPlayerType": {
            "location": "global"
          }
        },
        "return": "Promise<MediaProviderAdapter<InternalPlayerType> | undefined>"
      },
      "docs": {
        "text": "Returns the current media provider's adapter. Shorthand for `getProvider().getAdapter()`.",
        "tags": [{
            "name": "internal",
            "text": undefined
          }]
      }
    },
    "play": {
      "complexType": {
        "signature": "() => Promise<void | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void | undefined>"
      },
      "docs": {
        "text": "Begins/resumes playback of the media. If this method is called programmatically before the user\nhas interacted with the player, the promise may be rejected subject to the browser's autoplay\npolicies.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "pause": {
      "complexType": {
        "signature": "() => Promise<void | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void | undefined>"
      },
      "docs": {
        "text": "Pauses playback of the media.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canPlay": {
      "complexType": {
        "signature": "(type: string) => Promise<boolean>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Determines whether the current provider recognizes, and can play the given type.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canAutoplay": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Determines whether the player can start playback of the current media automatically.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canMutedAutoplay": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Determines whether the player can start playback of the current media automatically given the\nplayer is muted.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetPlaybackRate": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current provider allows setting the `playbackRate` prop.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetPlaybackQuality": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current provider allows setting the `playbackQuality` prop.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetFullscreen": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the native browser fullscreen API is available, or the current provider can\ntoggle fullscreen mode. This does not mean that the operation is guaranteed to be successful,\nonly that it can be attempted.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "enterFullscreen": {
      "complexType": {
        "signature": "(options?: FullscreenOptions | undefined) => Promise<any>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "FullscreenOptions": {
            "location": "global"
          }
        },
        "return": "Promise<any>"
      },
      "docs": {
        "text": "Requests to enter fullscreen mode, returning a `Promise` that will resolve if the request is\nmade, or reject with a reason for failure. This method will first attempt to use the browsers\nnative fullscreen API, and then fallback to requesting the provider to do it (if available).\nDo not rely on a resolved promise to determine if the player is in fullscreen or not. The only\nway to be certain is by listening to the `vmFullscreenChange` event. Some common reasons for\nfailure are: the fullscreen API is not available, the request is made when `viewType` is audio,\nor the user has not interacted with the page yet.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "exitFullscreen": {
      "complexType": {
        "signature": "() => Promise<any>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<any>"
      },
      "docs": {
        "text": "Requests to exit fullscreen mode, returning a `Promise` that will resolve if the request is\nsuccessful, or reject with a reason for failure. Refer to `enterFullscreen()` for more\ninformation.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetPiP": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current provider exposes an API for entering and exiting\npicture-in-picture mode. This does not mean the operation is guaranteed to be successful, only\nthat it can be attempted.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "enterPiP": {
      "complexType": {
        "signature": "() => Promise<void | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void | undefined>"
      },
      "docs": {
        "text": "Request to enter picture-in-picture (PiP) mode, returning a `Promise` that will resolve if\nthe request is made, or reject with a reason for failure. Do not rely on a resolved promise\nto determine if the player is in PiP mode or not. The only way to be certain is by listening\nto the `vmPiPChange` event. Some common reasons for failure are the same as the reasons for\n`enterFullscreen()`.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "exitPiP": {
      "complexType": {
        "signature": "() => Promise<void | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void | undefined>"
      },
      "docs": {
        "text": "Request to exit picture-in-picture mode, returns a `Promise` that will resolve if the request\nis successful, or reject with a reason for failure. Refer to `enterPiP()` for more\ninformation.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetAudioTrack": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current providers allows changing the audio track.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "setCurrentAudioTrack": {
      "complexType": {
        "signature": "(trackId: number) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Sets the currently active audio track given the index.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetTextTrack": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current provider allows changing the text track.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "setCurrentTextTrack": {
      "complexType": {
        "signature": "(trackId: number) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Sets the currently active text track given the index. Set to -1 to disable all text tracks.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "canSetTextTrackVisibility": {
      "complexType": {
        "signature": "() => Promise<boolean>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<boolean>"
      },
      "docs": {
        "text": "Returns whether the current providers allows setting the text track visibility.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "setTextTrackVisibility": {
      "complexType": {
        "signature": "(isVisible: boolean) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Sets the visibility of the currently active text track.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "extendLanguage": {
      "complexType": {
        "signature": "(language: string, translation: Partial<Translation>) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "Partial": {
            "location": "global"
          },
          "Translation": {
            "location": "import",
            "path": "./lang/Translation"
          },
          "Record": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Extends the translation map for a given language.",
        "tags": [{
            "name": "inheritDoc",
            "text": undefined
          }]
      }
    },
    "getContainer": {
      "complexType": {
        "signature": "() => Promise<HTMLDivElement | undefined>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          },
          "HTMLDivElement": {
            "location": "global"
          }
        },
        "return": "Promise<HTMLDivElement | undefined>"
      },
      "docs": {
        "text": "Returns the inner container.",
        "tags": []
      }
    },
    "callAdapter": {
      "complexType": {
        "signature": "(method: keyof MediaProviderAdapter, value?: any) => Promise<any>",
        "parameters": [{
            "tags": [],
            "text": ""
          }, {
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          },
          "MediaProviderAdapter": {
            "location": "import",
            "path": "../../providers/MediaProvider"
          }
        },
        "return": "Promise<any>"
      },
      "docs": {
        "text": "",
        "tags": [{
            "name": "internal",
            "text": "Exposed for E2E testing."
          }]
      }
    }
  }; }
  static get elementRef() { return "host"; }
  static get watchers() { return [{
      "propName": "container",
      "methodName": "onContainerChange"
    }, {
      "propName": "paused",
      "methodName": "onPausedChange"
    }, {
      "propName": "duration",
      "methodName": "onDurationChange"
    }, {
      "propName": "currentTime",
      "methodName": "onCurrentTimeChange"
    }, {
      "propName": "playbackReady",
      "methodName": "onPlaybackReadyChange"
    }, {
      "propName": "muted",
      "methodName": "onMutedChange"
    }, {
      "propName": "playbackRate",
      "methodName": "onPlaybackRateChange"
    }, {
      "propName": "playbackQuality",
      "methodName": "onPlaybackQualityChange"
    }, {
      "propName": "debug",
      "methodName": "onDebugChange"
    }, {
      "propName": "volume",
      "methodName": "onVolumeChange"
    }, {
      "propName": "viewType",
      "methodName": "onViewTypeChange"
    }, {
      "propName": "isAudioView",
      "methodName": "onViewTypeChange"
    }, {
      "propName": "isVideoView",
      "methodName": "onViewTypeChange"
    }, {
      "propName": "mediaType",
      "methodName": "onMediaTypeChange"
    }, {
      "propName": "language",
      "methodName": "onLanguageChange"
    }, {
      "propName": "translations",
      "methodName": "onTranslationsChange"
    }]; }
  static get listeners() { return [{
      "name": "vmError",
      "method": "onError",
      "target": undefined,
      "capture": false,
      "passive": false
    }]; }
}
