// javascript-time-ago/commonjs/PropTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.style = void 0;

var _propTypes = require("prop-types");

// Deprecated: Moved to `react-time-ago`.
// The first step isn't required to define `minTime` or `test()`.
var step = (0, _propTypes.oneOfType)([(0, _propTypes.shape)({
  minTime: _propTypes.number,
  formatAs: _propTypes.string.isRequired
}), (0, _propTypes.shape)({
  test: _propTypes.func,
  formatAs: _propTypes.string.isRequired
}), (0, _propTypes.shape)({
  minTime: _propTypes.number,
  format: _propTypes.func.isRequired
}), (0, _propTypes.shape)({
  test: _propTypes.func,
  format: _propTypes.func.isRequired
})]); // Formatting style.

var style = (0, _propTypes.oneOfType)([// Not using `oneOf()` here, because that way
// this package wouldn't support some hypothetical
// new styles added to `javascript-time-ago` in some future.
_propTypes.string, (0, _propTypes.shape)({
  steps: (0, _propTypes.arrayOf)(step).isRequired,
  labels: (0, _propTypes.oneOfType)([_propTypes.string, (0, _propTypes.arrayOf)(_propTypes.string)]).isRequired
})]);
exports.style = style;
