// devextreme/renovation/ui/pager/common/consts.js
/**
 * DevExtreme (renovation/ui/pager/common/consts.js)
 * Version: 20.2.6
 * Build date: Tue Mar 16 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.PAGER_SELECTED_PAGE_SIZE_CLASS = exports.PAGER_PAGE_SIZES_CLASS = exports.PAGER_PAGE_SIZE_CLASS = exports.PAGER_SELECTION_CLASS = exports.PAGER_PAGE_CLASS = exports.PAGER_PAGES_CLASS = exports.LIGHT_MODE_CLASS = exports.PAGER_CLASS = void 0;
var PAGER_CLASS = "dx-pager";
exports.PAGER_CLASS = PAGER_CLASS;
var LIGHT_MODE_CLASS = "dx-light-mode";
exports.LIGHT_MODE_CLASS = LIGHT_MODE_CLASS;
var PAGER_PAGES_CLASS = "dx-pages";
exports.PAGER_PAGES_CLASS = PAGER_PAGES_CLASS;
var PAGER_PAGE_CLASS = "dx-page";
exports.PAGER_PAGE_CLASS = PAGER_PAGE_CLASS;
var PAGER_SELECTION_CLASS = "dx-selection";
exports.PAGER_SELECTION_CLASS = PAGER_SELECTION_CLASS;
var PAGER_PAGE_SIZE_CLASS = "dx-page-size";
exports.PAGER_PAGE_SIZE_CLASS = PAGER_PAGE_SIZE_CLASS;
var PAGER_PAGE_SIZES_CLASS = "dx-page-sizes";
exports.PAGER_PAGE_SIZES_CLASS = PAGER_PAGE_SIZES_CLASS;
var PAGER_SELECTED_PAGE_SIZE_CLASS = "".concat(PAGER_PAGE_SIZE_CLASS, " ").concat(PAGER_SELECTION_CLASS);
exports.PAGER_SELECTED_PAGE_SIZE_CLASS = PAGER_SELECTED_PAGE_SIZE_CLASS;
