// @progress/kendo-react-labels/dist/npm/Error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_licensing_1 = require("@progress/kendo-licensing");
var package_metadata_1 = require("./package-metadata");
/**
 * Represents the KendoReact Error component.
 * Render the error text that will be shown underneath the form editor after a validation.
 */
exports.Error = function (props) {
    kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
    var errorClassName = kendo_react_common_1.classNames({
        'k-form-error': true,
        'k-text-start': props.direction === 'start',
        'k-text-end': props.direction === 'end'
    }, props.className);
    return (React.createElement("div", { id: props.id, role: 'alert', style: props.style, className: errorClassName }, props.children));
};
exports.Error.propTypes = {
    id: PropTypes.string,
    direction: PropTypes.oneOf(['start', 'end']),
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.node
    ]),
    style: PropTypes.object,
    className: PropTypes.string
};
exports.Error.defaultProps = {
    direction: 'start'
};
exports.Error.displayName = 'KendoReactError';
