// @progress/kendo-react-labels/dist/npm/Hint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_licensing_1 = require("@progress/kendo-licensing");
var package_metadata_1 = require("./package-metadata");
/**
 * Represents the KendoReact Hint component.
 * Render the hint text that will be shown underneath the form editor.
 */
exports.Hint = function (props) {
    kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
    var hintClassName = kendo_react_common_1.classNames({
        'k-form-hint': true,
        'k-text-start': props.direction === 'start',
        'k-text-end': props.direction === 'end'
    }, props.className);
    return (React.createElement("div", { id: props.id, style: props.style, className: hintClassName }, props.children));
};
exports.Hint.propTypes = {
    id: PropTypes.string,
    direction: PropTypes.oneOf(['start', 'end']),
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.node
    ]),
    style: PropTypes.object,
    className: PropTypes.string
};
exports.Hint.defaultProps = {
    direction: 'start'
};
exports.Hint.displayName = 'KendoReactHint';
